// @openai/agents-core/dist/sandbox/runtime/spans.mjs
import { getCurrentTrace, withCustomSpan } from "../../tracing/index.mjs";
import { emitSandboxEvent, serializeSandboxEventError } from "../events.mjs";
export async function withSandboxSpan(name, data, fn) {
    const startedAt = Date.now();
    await emitSandboxEvent({
        type: 'sandbox_operation',
        name,
        phase: 'start',
        timestamp: new Date(startedAt).toISOString(),
        data: { ...data },
    });
    const runWithEvents = async () => {
        try {
            const result = await fn();
            await emitSandboxEvent({
                type: 'sandbox_operation',
                name,
                phase: 'end',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
            });
            return result;
        }
        catch (error) {
            await emitSandboxEvent({
                type: 'sandbox_operation',
                name,
                phase: 'error',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
                error: serializeSandboxEventError(error),
            });
            throw error;
        }
    };
    if (!getCurrentTrace()) {
        return await runWithEvents();
    }
    return await withCustomSpan(async () => await runWithEvents(), {
        data: {
            name,
            data,
        },
    });
}
