// @openai/agents-core/dist/sandbox/runtime/spans.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withSandboxSpan = withSandboxSpan;
const tracing_1 = require("../../tracing/index.js");
const events_1 = require("../events.js");
async function withSandboxSpan(name, data, fn) {
    const startedAt = Date.now();
    await (0, events_1.emitSandboxEvent)({
        type: 'sandbox_operation',
        name,
        phase: 'start',
        timestamp: new Date(startedAt).toISOString(),
        data: { ...data },
    });
    const runWithEvents = async () => {
        try {
            const result = await fn();
            await (0, events_1.emitSandboxEvent)({
                type: 'sandbox_operation',
                name,
                phase: 'end',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
            });
            return result;
        }
        catch (error) {
            await (0, events_1.emitSandboxEvent)({
                type: 'sandbox_operation',
                name,
                phase: 'error',
                timestamp: new Date().toISOString(),
                data: { ...data },
                durationMs: Date.now() - startedAt,
                error: (0, events_1.serializeSandboxEventError)(error),
            });
            throw error;
        }
    };
    if (!(0, tracing_1.getCurrentTrace)()) {
        return await runWithEvents();
    }
    return await (0, tracing_1.withCustomSpan)(async () => await runWithEvents(), {
        data: {
            name,
            data,
        },
    });
}
