// @corsair-dev/mcp/dist/index.js
// src/adapters/anthropic-api.ts
import { z as z3 } from "zod";

// src/core/json-schema.ts
import { z } from "zod";
import * as z4mini from "zod/v4-mini";
function shapeToJsonSchema(shape) {
  return z4mini.toJSONSchema(z.object(shape), {
    target: "draft-7",
    io: "input"
  });
}

// src/core/tools.ts
import { listOperations, setupCorsair } from "corsair";
import { z as z2 } from "zod";

// src/core/schema-format.ts
import { formatDocSchemaShape, getSchema } from "corsair";
function formatGetSchemaResponse(corsair, path) {
  return getSchema(corsair, path);
}

// src/core/tools.ts
function buildCorsairToolDefs(options) {
  const {
    corsair,
    permissions,
    basePermissionUrl,
    setup,
    tenantId: defaultTenantId
  } = options;
  const defs = [
    {
      name: "list_operations",
      description: "List available Corsair operations. Without options returns all API endpoints across every plugin. Filter by plugin (e.g. 'slack') and/or type ('api' | 'webhooks' | 'db').",
      shape: {
        plugin: z2.string().optional().describe("Plugin ID to filter by, e.g. 'slack' or 'github'"),
        type: z2.enum(["api", "webhooks", "db"]).optional().describe("Operation type: 'api' (default), 'webhooks', or 'db'")
      },
      handler: async ({ plugin, type }) => {
        const result = listOperations(corsair, {
          plugin,
          type
        });
        return {
          content: [{ type: "text", text: result }]
        };
      }
    },
    {
      name: "get_schema",
      description: "Get the schema and metadata for a Corsair operation path. Accepts API paths ('slack.api.channels.list'), webhook paths ('slack.webhooks.messages.message'), or DB paths ('slack.db.messages.search').",
      shape: {
        path: z2.string().describe(
          "Full dot-path from list_operations, e.g. 'slack.api.channels.list'"
        )
      },
      handler: async ({ path }) => {
        const result = formatGetSchemaResponse(
          corsair,
          path
        );
        return {
          content: [{ type: "text", text: result }]
        };
      }
    },
    {
      name: "run_script",
      description: "Run a JavaScript script with `corsair` as the only variable in scope. Call Corsair operations, filter or transform the results inline, and return only what you need. The return value becomes the tool output.",
      shape: {
        code: z2.string().describe(
          'Async JS script with `corsair` in scope. Return the value you want. Example:\nconst result = await corsair.slack.api.channels.list({});\nconst channel = result.channels?.find(c => c.name === "general");\nreturn channel?.id;'
        )
      },
      handler: async ({ code }) => {
        try {
          const fn = new Function(
            "corsair",
            `return (async () => { ${code} })()`
          );
          const result = await fn(
            corsair
          );
          return {
            content: [
              {
                type: "text",
                text: JSON.stringify(result ?? null, null, 2)
              }
            ]
          };
        } catch (err) {
          const message = err instanceof Error ? err.message : String(err);
          const extra = err instanceof Error && err.cause ? `
Cause: ${String(err.cause)}` : "";
          const full = JSON.stringify(err, Object.getOwnPropertyNames(err));
          return {
            isError: true,
            content: [
              {
                type: "text",
                text: `Error running snippet: ${message}${extra}
${full}`
              }
            ]
          };
        }
      }
    }
  ];
  if (setup == null || setup === true) {
    defs.push({
      name: "corsair_setup",
      description: "Helps the user configure Corsair. Call this to see if any keys or tokens need to be set up. It will also provide the instructions to set them up. For multi-tenant Corsair instances, pass tenantId to set up a specific tenant.",
      shape: {
        tenantId: z2.string().optional().describe(
          "Tenant ID to configure for multi-tenant Corsair instances. Defaults to the server's configured tenantId, then 'default'."
        )
      },
      handler: async ({ tenantId }) => {
        try {
          const setupTenantId = typeof tenantId === "string" && tenantId.length > 0 ? tenantId : defaultTenantId;
          const text = await setupCorsair(
            corsair,
            setupTenantId ? { tenantId: setupTenantId } : void 0
          );
          return {
            content: [
              { type: "text", text: text || "Corsair setup complete." }
            ]
          };
        } catch (err) {
          const message = err instanceof Error ? err.message : String(err);
          return {
            isError: true,
            content: [{ type: "text", text: `Setup failed: ${message}` }]
          };
        }
      }
    });
  }
  if (permissions && basePermissionUrl) {
    defs.push({
      name: "request_permission",
      description: "Request permission from the user to execute a protected endpoint. Call this when a script returns a [PERMISSION_REQUIRED] message. Returns an approval URL. After calling this, call ask_human with the approval URL so the user can review and approve.",
      shape: {
        endpoint: z2.string().describe(
          'Full endpoint path from the PERMISSION_REQUIRED message, e.g. "slack.messages.post"'
        ),
        args: z2.record(z2.string(), z2.unknown()).describe(
          "The arguments object from the PERMISSION_REQUIRED message"
        ),
        description: z2.string().describe(
          'Short human-readable summary of what this action will do, e.g. "Post a message to #general in Slack"'
        ),
        jid: z2.string().optional().describe(
          "Optional chat/conversation ID for resuming after approval"
        )
      },
      handler: async ({ endpoint, args, description }) => {
        const { permissionId, approvalUrl } = await permissions.createPermissionRequest({
          endpoint,
          args,
          description
        });
        const result = {
          permissionId,
          approvalUrl,
          message: `Permission request created. Ask the user to approve at: ${approvalUrl}`
        };
        return {
          content: [{ type: "text", text: JSON.stringify(result) }]
        };
      }
    });
  }
  return defs;
}

// src/core/provider.ts
var BaseProvider = class {
  build(options) {
    return buildCorsairToolDefs(options).map((def) => this.wrapTool(def));
  }
};

// src/adapters/anthropic-api.ts
var AnthropicProvider = class extends BaseProvider {
  name = "anthropic";
  wrapTool(def) {
    const schema = shapeToJsonSchema(def.shape);
    return {
      type: "custom",
      name: def.name,
      description: def.description,
      input_schema: schema,
      run: async (args) => {
        try {
          const result = await def.handler(args);
          return result.content.filter((c) => c.type === "text").map((c) => "text" in c ? c.text : "").join("\n");
        } catch (err) {
          const message = err instanceof Error ? err.message : String(err);
          return `Error: ${message}`;
        }
      },
      parse: (content) => z3.object(def.shape).parse(content)
    };
  }
};

// src/adapters/claude.ts
var ClaudeProvider = class {
  name = "claude";
  async build(options) {
    const { tool } = await import(
      /* webpackIgnore: true */
      /* turbopackIgnore: true */
      "@anthropic-ai/claude-agent-sdk"
    );
    return buildCorsairToolDefs(options).map(
      (def) => tool(def.name, def.description, def.shape, def.handler)
    );
  }
};

// src/adapters/mastra.ts
import { z as z4 } from "zod";
var MastraProvider = class {
  name = "mastra";
  async build(options) {
    const { createTool } = await import(
      /* webpackIgnore: true */
      /* turbopackIgnore: true */
      "@mastra/core/tools"
    );
    return buildCorsairToolDefs(options).map(
      (def) => createTool({
        id: def.name,
        description: def.description,
        inputSchema: z4.object(def.shape),
        execute: async (inputData) => {
          const result = await def.handler(
            inputData
          );
          const textItems = result.content.filter((c) => c.type === "text");
          if (result.isError) {
            throw new Error(textItems.map((c) => c.text).join("\n"));
          }
          const text = textItems[0]?.text ?? "";
          try {
            return JSON.parse(text);
          } catch {
            return text;
          }
        }
      })
    );
  }
};

// src/adapters/ollama.ts
import { z as z5 } from "zod";
var OllamaProvider = class extends BaseProvider {
  name = "ollama";
  wrapTool(def) {
    const schema = shapeToJsonSchema(def.shape);
    return {
      type: "function",
      function: {
        name: def.name,
        description: def.description,
        parameters: schema
      },
      /**
       * The `execute` helper provides a simplified execution interface for runners.
       * It handles:
       * 1. Parsing arguments (handles both object and JSON string inputs).
       * 2. Validation against the tool's Zod schema.
       * 3. Execution of the underlying Corsair tool handler.
       * 4. Extraction and concatenation of text results into a single string.
       * 5. Graceful error handling for robust integration.
       */
      execute: async (args) => {
        try {
          const raw = typeof args === "string" ? JSON.parse(args) : args;
          const validated = z5.object(def.shape).parse(raw);
          const result = await def.handler(validated);
          return result.content.filter((c) => c.type === "text").map((c) => "text" in c ? c.text : "").join("\n");
        } catch (err) {
          const message = err instanceof Error ? err.message : String(err);
          return `Error: ${message}`;
        }
      }
    };
  }
};

// src/adapters/openai.ts
function getOpenAIMcpConfig(baseUrl, headers) {
  return {
    serverLabel: "corsair",
    serverUrl: baseUrl,
    headers
  };
}

// src/adapters/openai-agents.ts
import { z as z6 } from "zod";
var OpenAIAgentsProvider = class {
  name = "openai-agents";
  build(options) {
    const { tool, ...mcpOptions } = options;
    return buildCorsairToolDefs(mcpOptions).map((def) => {
      const jsonSchema = shapeToJsonSchema(def.shape);
      const shapeKeys = new Set(Object.keys(def.shape));
      return tool({
        name: def.name,
        description: def.description,
        parameters: jsonSchema,
        strict: false,
        execute: async (input) => {
          try {
            const raw = typeof input === "string" ? JSON.parse(input) : input;
            if (shapeKeys.has("args") && typeof raw.args !== "object") {
              const extra = {};
              for (const [k, v] of Object.entries(raw)) {
                if (!shapeKeys.has(k)) extra[k] = v;
              }
              if (Object.keys(extra).length > 0) raw.args = extra;
            }
            const args = z6.object(def.shape).parse(raw);
            const result = await def.handler(args);
            return result.content.filter((c) => c.type === "text").map((c) => "text" in c ? c.text : "").join("\n");
          } catch (err) {
            const message = err instanceof Error ? err.message : String(err);
            return `Error: ${message}`;
          }
        }
      });
    });
  }
};

// src/adapters/vercel-ai.ts
async function createVercelAiMcpClient(options) {
  const { createMCPClient } = await import(
    /* webpackIgnore: true */
    /* turbopackIgnore: true */
    "@ai-sdk/mcp"
  );
  const { url, headers } = options;
  const client = await createMCPClient({
    transport: {
      type: "http",
      url,
      headers
    }
  });
  return client;
}

// src/core/base.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z as z7 } from "zod";
function createBaseMcpServer(options) {
  const server = new McpServer({
    name: "corsair",
    version: "1.0.0",
    description: "Use this to interact with the Corsair API. Corsair helps you integrate with dozens of tools and services."
  });
  for (const def of buildCorsairToolDefs(options)) {
    server.registerTool(
      def.name,
      { description: def.description, inputSchema: z7.object(def.shape) },
      def.handler
    );
  }
  return server;
}

// src/core/http.ts
import { StreamableHTTPServerTransport } from "@modelcontextprotocol/sdk/server/streamableHttp.js";
import { randomUUID } from "crypto";
import { Router as createRouter } from "express";
var sessions = /* @__PURE__ */ new Map();
function cleanup(sessionId) {
  const session = sessions.get(sessionId);
  if (session) {
    session.transport.close();
    session.server.close();
    sessions.delete(sessionId);
  }
}
function createMcpRouter(createServer) {
  const router = createRouter();
  router.post("/", async (req, res) => {
    const sessionId = req.headers["mcp-session-id"];
    if (sessionId) {
      const session = sessions.get(sessionId);
      if (!session) {
        res.status(404).json({ error: "Session not found" });
        return;
      }
      await session.transport.handleRequest(req, res, req.body);
      return;
    }
    const server = createServer();
    const transport = new StreamableHTTPServerTransport({
      sessionIdGenerator: () => randomUUID(),
      onsessioninitialized: (id) => {
        sessions.set(id, { server, transport });
      }
    });
    res.on("close", () => {
      const id = transport.sessionId;
      if (id) {
        setTimeout(() => cleanup(id), 6e4);
      }
    });
    await server.connect(transport);
    await transport.handleRequest(req, res, req.body);
  });
  router.get("/", async (req, res) => {
    const sessionId = req.headers["mcp-session-id"];
    if (!sessionId || !sessions.has(sessionId)) {
      res.status(400).json({ error: "Missing or invalid mcp-session-id" });
      return;
    }
    const session = sessions.get(sessionId);
    await session.transport.handleRequest(req, res);
  });
  router.delete("/", async (req, res) => {
    const sessionId = req.headers["mcp-session-id"];
    if (sessionId) cleanup(sessionId);
    res.status(200).end();
  });
  return router;
}

// src/core/stdio.ts
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
async function runStdioMcpServer(options) {
  const server = createBaseMcpServer(options);
  const transport = new StdioServerTransport();
  await server.connect(transport);
}
export {
  AnthropicProvider,
  BaseProvider,
  ClaudeProvider,
  MastraProvider,
  OllamaProvider,
  OpenAIAgentsProvider,
  buildCorsairToolDefs,
  createBaseMcpServer,
  createMcpRouter,
  createVercelAiMcpClient,
  getOpenAIMcpConfig,
  runStdioMcpServer
};
