// @corsair-dev/cli/dist/index.js
#!/usr/bin/env node
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __esm = (fn, res) => function __init() {
  return fn && (res = (0, fn[__getOwnPropNames(fn)[0]])(fn = 0)), res;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/get-tsconfig-info.ts
import fs from "fs";
import path from "path";
function stripJsonComments(jsonString) {
  return jsonString.replace(
    /\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g,
    (m, g) => g ? "" : m
  ).replace(/,(?=\s*[}\]])/g, "");
}
function getTsconfigInfo(cwd, flatPath) {
  let tsConfigPath;
  if (flatPath) {
    tsConfigPath = flatPath;
  } else {
    tsConfigPath = cwd ? path.join(cwd, "tsconfig.json") : path.join("tsconfig.json");
  }
  try {
    const text6 = fs.readFileSync(tsConfigPath, "utf-8");
    return JSON.parse(stripJsonComments(text6));
  } catch (error) {
    throw error;
  }
}
var init_get_tsconfig_info = __esm({
  "src/get-tsconfig-info.ts"() {
    "use strict";
  }
});

// src/utils/corsair-instance.ts
import fs2, { existsSync } from "fs";
import path2 from "path";
import babelPresetReact from "@babel/preset-react";
import babelPresetTypeScript from "@babel/preset-typescript";
import { loadConfig } from "c12";
function isCorsairInstance(value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return false;
  }
  if ("withTenant" in value && typeof value.withTenant === "function") {
    return true;
  }
  return Object.keys(value).length > 0;
}
function resolveReferencePath(configDir, refPath) {
  const resolvedPath = path2.resolve(configDir, refPath);
  if (refPath.endsWith(".json")) return resolvedPath;
  if (existsSync(resolvedPath)) {
    const stats = fs2.statSync(resolvedPath);
    if (stats.isFile()) return resolvedPath;
  }
  return path2.resolve(configDir, refPath, "tsconfig.json");
}
function getPathAliasesRecursive(tsconfigPath, visited = /* @__PURE__ */ new Set()) {
  if (visited.has(tsconfigPath)) return {};
  visited.add(tsconfigPath);
  if (!existsSync(tsconfigPath)) return {};
  const tsConfig = getTsconfigInfo(void 0, tsconfigPath);
  const { paths = {}, baseUrl = "." } = tsConfig.compilerOptions || {};
  const configDir = path2.dirname(tsconfigPath);
  const resolvedBaseUrl = path2.resolve(configDir, baseUrl);
  const aliases = {};
  for (const [alias, aliasPaths] of Object.entries(paths)) {
    for (const aliasPath of aliasPaths) {
      const aliasKey = alias.endsWith("*") ? alias.slice(0, -1) : alias;
      const aliasValue = aliasPath.endsWith("*") ? aliasPath.slice(0, -1) : aliasPath;
      aliases[aliasKey] = path2.join(resolvedBaseUrl, aliasValue);
    }
  }
  for (const ref of tsConfig.references || []) {
    const refPath = resolveReferencePath(configDir, ref.path);
    const refAliases = getPathAliasesRecursive(refPath, visited);
    for (const [alias, aliasPath] of Object.entries(refAliases)) {
      if (!(alias in aliases)) aliases[alias] = aliasPath;
    }
  }
  return aliases;
}
function getPathAliases(cwd) {
  const tsConfigPath = path2.join(cwd, "tsconfig.json");
  return existsSync(tsConfigPath) ? getPathAliasesRecursive(tsConfigPath) : {};
}
function jitiOptions(cwd) {
  return {
    transformOptions: {
      babel: {
        presets: [
          [babelPresetTypeScript, { isTSX: true, allExtensions: true }],
          [babelPresetReact, { runtime: "automatic" }]
        ]
      }
    },
    extensions: [".ts", ".tsx", ".js", ".jsx"],
    alias: getPathAliases(cwd)
  };
}
function fail(message, shouldThrowOnError) {
  if (shouldThrowOnError) throw new Error(message);
  console.error(`[#corsair]: ${message}`);
  process.exit(1);
}
function parseLoadError(error) {
  const msg = error instanceof Error ? error.message : String(error);
  if (msg.includes(
    "This module cannot be imported from a Client Component module"
  )) {
    return "Please remove import 'server-only' from your corsair.ts file temporarily.";
  }
  if (msg.includes("Could not locate the bindings file") || msg.includes("NODE_MODULE_VERSION") || msg.includes(".node")) {
    return "Native module binding not found. Run `npm rebuild` or reinstall dependencies.";
  }
  return msg;
}
function findCorsairConfigPath(cwd) {
  for (const possiblePath of POSSIBLE_PATHS) {
    const fullPath = path2.join(cwd, possiblePath);
    if (existsSync(fullPath)) return fullPath;
  }
  return null;
}
async function getCorsairInstance({
  cwd,
  configPath,
  shouldThrowOnError = false
}) {
  const candidatePaths = configPath ? [existsSync(configPath) ? configPath : path2.join(cwd, configPath)] : [...POSSIBLE_PATHS];
  for (const candidatePath of candidatePaths) {
    try {
      const { config } = await loadConfig({
        configFile: candidatePath,
        dotenv: true,
        jitiOptions: jitiOptions(cwd),
        cwd
      });
      if ("corsair" in config && isCorsairInstance(config.corsair)) {
        return config.corsair;
      }
      if ("default" in config && isCorsairInstance(config.default)) {
        return config.default;
      }
      if (Object.keys(config).length > 0) {
        fail(
          `Couldn't read your Corsair instance in ${candidatePath}. Export it as 'corsair' or default.`,
          shouldThrowOnError
        );
      }
    } catch (error) {
      fail(
        `Error loading ${candidatePath}: ${parseLoadError(error)}`,
        shouldThrowOnError
      );
    }
  }
  fail(
    "Couldn't find corsair.ts in your project. Add it in root or src/lib/server/app.",
    shouldThrowOnError
  );
}
function resolveClient(instance, tenant) {
  const obj = instance;
  if ("withTenant" in obj && typeof obj.withTenant === "function") {
    if (!tenant) {
      console.error(
        "[#corsair]: This is a multi-tenant instance. Pass --tenant=<id>."
      );
      process.exit(1);
    }
    return obj.withTenant(tenant);
  }
  return obj;
}
var POSSIBLE_PATHS;
var init_corsair_instance = __esm({
  "src/utils/corsair-instance.ts"() {
    "use strict";
    init_get_tsconfig_info();
    POSSIBLE_PATHS = [
      "corsair.ts",
      "corsair.tsx",
      "corsair.js",
      "corsair.jsx",
      "src/corsair.ts",
      "src/corsair.tsx",
      "src/corsair.js",
      "src/corsair.jsx",
      "lib/corsair.ts",
      "lib/corsair.tsx",
      "lib/corsair.js",
      "lib/corsair.jsx",
      "server/corsair.ts",
      "server/corsair.tsx",
      "server/corsair.js",
      "server/corsair.jsx",
      "src/server/corsair.ts",
      "src/server/corsair.tsx",
      "src/server/corsair.js",
      "src/server/corsair.jsx",
      "app/corsair.ts",
      "app/corsair.tsx",
      "app/corsair.js",
      "app/corsair.jsx",
      "corsair/index.ts",
      "corsair/index.tsx",
      "corsair/index.js",
      "corsair/index.jsx"
    ];
  }
});

// src/utils/corsair.ts
import { CORSAIR_INTERNAL } from "corsair/core";
async function extractInternalConfig(cwd) {
  const instance = await getCorsairInstance({ cwd, shouldThrowOnError: true });
  const internal = instance[CORSAIR_INTERNAL];
  if (!internal) {
    throw new Error(
      "Could not read internal config from Corsair instance. Make sure you are using the latest version of corsair."
    );
  }
  return internal;
}
var init_corsair = __esm({
  "src/utils/corsair.ts"() {
    "use strict";
    init_corsair_instance();
  }
});

// src/utils/load-config.ts
import * as p from "@clack/prompts";
async function loadInternalConfig(cwd, title, pluginId, pluginLabel) {
  p.intro(title);
  const spin = p.spinner();
  spin.start("Loading corsair instance...");
  let internal;
  try {
    internal = await extractInternalConfig(cwd);
  } catch (error) {
    spin.stop("Failed to load.");
    p.log.error(error instanceof Error ? error.message : String(error));
    p.outro("");
    process.exit(1);
  }
  const { plugins, database } = internal;
  if (!database) {
    spin.stop("Failed.");
    p.log.error("No database adapter configured.");
    p.outro("");
    process.exit(1);
  }
  const plugin = plugins.find((pl) => pl.id === pluginId);
  if (!plugin) {
    spin.stop(`${pluginLabel} plugin not found.`);
    p.log.error(`Add the ${pluginId} plugin to your corsair instance first.`);
    p.outro("");
    process.exit(1);
  }
  spin.stop("Loaded.");
  return { internal, plugin };
}
var init_load_config = __esm({
  "src/utils/load-config.ts"() {
    "use strict";
    init_corsair();
  }
});

// src/utils/prompts.ts
import * as crypto from "crypto";
import * as p2 from "@clack/prompts";
async function promptTenantId() {
  const tenantId = await p2.text({
    message: "Enter tenant ID:",
    defaultValue: "default",
    placeholder: "default"
  });
  if (p2.isCancel(tenantId)) {
    p2.cancel("Operation cancelled.");
    process.exit(0);
  }
  return tenantId;
}
async function promptWebhookUrl() {
  const webhookUrl = await p2.text({
    message: "Enter your public webhook URL (ngrok or production):",
    placeholder: "https://abc123.ngrok-free.app/api/webhook",
    validate: (v) => {
      if (!v || v.trim().length === 0) return "Webhook URL is required";
      if (!v.startsWith("https://"))
        return "URL must start with https:// (Microsoft Graph requires HTTPS)";
    }
  });
  if (p2.isCancel(webhookUrl)) {
    p2.cancel("Operation cancelled.");
    process.exit(0);
  }
  return webhookUrl;
}
async function promptClientState() {
  const autoSecret = crypto.randomBytes(16).toString("hex");
  const clientState = await p2.text({
    message: "Enter a clientState secret (used to verify webhook payloads):",
    defaultValue: autoSecret,
    placeholder: autoSecret
  });
  if (p2.isCancel(clientState)) {
    p2.cancel("Operation cancelled.");
    process.exit(0);
  }
  return clientState;
}
var init_prompts = __esm({
  "src/utils/prompts.ts"() {
    "use strict";
  }
});

// src/lib/microsoft/credentials.ts
import * as p3 from "@clack/prompts";
import {
  createAccountKeyManager,
  createIntegrationKeyManager
} from "corsair/core";
async function refreshMicrosoftToken(clientId, clientSecret, refreshToken) {
  const response = await fetch(MICROSOFT_TOKEN_URL, {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: clientId,
      client_secret: clientSecret,
      refresh_token: refreshToken,
      grant_type: "refresh_token"
    })
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Token refresh failed: ${error}`);
  }
  return response.json();
}
async function resolveAccessToken(integrationName, tenantId, internal) {
  const { kek, database } = internal;
  const credSpin = p3.spinner();
  credSpin.start(`Fetching ${integrationName} credentials...`);
  const integrationKm = createIntegrationKeyManager({
    authType: "oauth_2",
    integrationName,
    kek,
    database
  });
  const accountKm = createAccountKeyManager({
    authType: "oauth_2",
    integrationName,
    tenantId,
    kek,
    database
  });
  const [
    clientId,
    clientSecret,
    storedAccessToken,
    expiresAt,
    storedRefreshToken
  ] = await Promise.all([
    integrationKm.get_client_id(),
    integrationKm.get_client_secret(),
    accountKm.get_access_token(),
    accountKm.get_expires_at(),
    accountKm.get_refresh_token()
  ]);
  if (!storedRefreshToken) {
    credSpin.stop("Missing credentials.");
    p3.log.error(
      `No refresh token found. Run: pnpm corsair auth --plugin=${integrationName}`
    );
    p3.outro("");
    process.exit(1);
  }
  if (!clientId || !clientSecret) {
    credSpin.stop("Missing credentials.");
    p3.log.error(
      `Client ID/Secret not configured. Run: pnpm corsair setup --${integrationName}`
    );
    p3.outro("");
    process.exit(1);
  }
  let accessToken = storedAccessToken;
  const now = Math.floor(Date.now() / 1e3);
  const needsRefresh = !accessToken || !expiresAt || Number(expiresAt) <= now + 5 * 60;
  if (needsRefresh) {
    credSpin.message("Refreshing access token...");
    try {
      const tokenData = await refreshMicrosoftToken(
        clientId,
        clientSecret,
        storedRefreshToken
      );
      accessToken = tokenData.access_token;
      await accountKm.set_access_token(tokenData.access_token);
      await accountKm.set_expires_at(String(now + tokenData.expires_in));
      if (tokenData.refresh_token) {
        await accountKm.set_refresh_token(tokenData.refresh_token);
      }
    } catch (error) {
      credSpin.stop("Failed.");
      p3.log.error(error instanceof Error ? error.message : String(error));
      p3.outro("");
      process.exit(1);
    }
  }
  if (!accessToken) {
    credSpin.stop("Missing credentials.");
    p3.log.error(
      `Could not obtain a valid access token. Run: pnpm corsair auth --plugin=${integrationName}`
    );
    p3.outro("");
    process.exit(1);
  }
  credSpin.stop("Credentials loaded.");
  return { accessToken, accountKm };
}
async function saveWebhookSignature(accountKm, clientState) {
  const saveSpin = p3.spinner();
  saveSpin.start("Saving webhook secret...");
  await accountKm.set_webhook_signature(clientState);
  saveSpin.stop("Webhook secret saved.");
}
var MICROSOFT_TOKEN_URL;
var init_credentials = __esm({
  "src/lib/microsoft/credentials.ts"() {
    "use strict";
    MICROSOFT_TOKEN_URL = "https://login.microsoftonline.com/common/oauth2/v2.0/token";
  }
});

// src/lib/microsoft/graph.ts
async function createGraphSubscription(accessToken, notificationUrl, resource, changeType, clientState, expiryMinutes) {
  const expirationDateTime = new Date(
    Date.now() + expiryMinutes * 60 * 1e3
  ).toISOString();
  const response = await fetch(`${GRAPH_API_BASE}/subscriptions`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      changeType,
      notificationUrl,
      resource,
      expirationDateTime,
      clientState
    })
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Graph subscription failed: ${error}`);
  }
  return response.json();
}
var GRAPH_API_BASE;
var init_graph = __esm({
  "src/lib/microsoft/graph.ts"() {
    "use strict";
    GRAPH_API_BASE = "https://graph.microsoft.com/v1.0";
  }
});

// src/lib/microsoft/onedrive.ts
import * as p4 from "@clack/prompts";
function buildOnedriveResource(resourceType, ids) {
  switch (resourceType) {
    case "personalDrive":
      return "me/drive/root";
    case "specificDrive":
      return `drives/${ids.driveId}/root`;
    case "specificFolder":
      return `me/drive/items/${ids.itemId}`;
  }
}
async function runOnedriveSubscribe({
  cwd
}) {
  const { internal } = await loadInternalConfig(
    cwd,
    "Corsair \u2014 OneDrive Webhook Subscribe",
    "onedrive",
    "OneDrive"
  );
  const tenantId = await promptTenantId();
  const { accessToken, accountKm } = await resolveAccessToken(
    "onedrive",
    tenantId,
    internal
  );
  const webhookUrl = await promptWebhookUrl();
  const selectedResourceTypes = await p4.multiselect({
    message: "Select resources to subscribe to:",
    options: ONEDRIVE_RESOURCE_OPTIONS,
    required: true
  });
  if (p4.isCancel(selectedResourceTypes)) {
    p4.cancel("Operation cancelled.");
    process.exit(0);
  }
  const resourceConfigs = [];
  for (const resourceType of selectedResourceTypes) {
    const ids = {};
    if (resourceType === "specificDrive") {
      const driveId = await p4.text({
        message: `[${resourceType}] Enter drive ID:`,
        placeholder: "b!abc123...",
        validate: (v) => {
          if (!v || !v.trim()) return "Drive ID is required";
        }
      });
      if (p4.isCancel(driveId)) {
        p4.cancel("Operation cancelled.");
        process.exit(0);
      }
      ids.driveId = driveId;
    } else if (resourceType === "specificFolder") {
      const itemId = await p4.text({
        message: `[${resourceType}] Enter item/folder ID:`,
        placeholder: "01ABC123...",
        validate: (v) => {
          if (!v || !v.trim()) return "Item ID is required";
        }
      });
      if (p4.isCancel(itemId)) {
        p4.cancel("Operation cancelled.");
        process.exit(0);
      }
      ids.itemId = itemId;
    }
    resourceConfigs.push({ resourceType, ids });
  }
  const clientState = await promptClientState();
  const results = [];
  let hasError = false;
  for (const { resourceType, ids } of resourceConfigs) {
    const resource = buildOnedriveResource(resourceType, ids);
    const subSpin = p4.spinner();
    subSpin.start(`Creating subscription: ${resourceType}...`);
    try {
      const subscription = await createGraphSubscription(
        accessToken,
        webhookUrl,
        resource,
        "updated",
        clientState,
        ONEDRIVE_MAX_EXPIRY_MINUTES
      );
      subSpin.stop(`Created: ${resourceType}`);
      results.push(
        `${resourceType}
  Resource: ${resource}
  ID: ${subscription.id}
  Expires: ${subscription.expirationDateTime}`
      );
    } catch (error) {
      subSpin.stop(`Failed: ${resourceType}`);
      p4.log.error(error instanceof Error ? error.message : String(error));
      hasError = true;
    }
  }
  if (results.length > 0) {
    p4.note(results.join("\n\n"), "Subscriptions created");
  }
  await saveWebhookSignature(accountKm, clientState);
  if (!hasError) {
    p4.log.success("OneDrive webhook subscriptions set up successfully.");
  } else {
    p4.log.warn("Some subscriptions failed. Check errors above.");
  }
  p4.note(
    `ClientState: ${clientState}
Webhook URL: ${webhookUrl}

Note: OneDrive subscriptions expire. Re-run this command to renew.`,
    "Setup complete"
  );
  p4.outro("Done!");
}
var ONEDRIVE_MAX_EXPIRY_MINUTES, ONEDRIVE_RESOURCE_OPTIONS;
var init_onedrive = __esm({
  "src/lib/microsoft/onedrive.ts"() {
    "use strict";
    init_load_config();
    init_prompts();
    init_credentials();
    init_graph();
    ONEDRIVE_MAX_EXPIRY_MINUTES = 4230;
    ONEDRIVE_RESOURCE_OPTIONS = [
      {
        value: "personalDrive",
        label: "Personal Drive Root \u2014 all changes to the user's drive",
        hint: "me/drive/root"
      },
      {
        value: "specificDrive",
        label: "Specific Drive Root \u2014 all changes to a given drive ID",
        hint: "drives/{driveId}/root"
      },
      {
        value: "specificFolder",
        label: "Specific Folder \u2014 changes within a given item/folder",
        hint: "me/drive/items/{itemId}"
      }
    ];
  }
});

// src/lib/microsoft/outlook.ts
import * as p5 from "@clack/prompts";
async function runOutlookSubscribe({
  cwd
}) {
  const { internal } = await loadInternalConfig(
    cwd,
    "Corsair \u2014 Outlook Webhook Subscribe",
    "outlook",
    "Outlook"
  );
  const tenantId = await promptTenantId();
  const { accessToken, accountKm } = await resolveAccessToken(
    "outlook",
    tenantId,
    internal
  );
  const webhookUrl = await promptWebhookUrl();
  const selected = await p5.multiselect({
    message: "Select resources to subscribe to:",
    options: Object.entries(OUTLOOK_RESOURCES).map(([value, cfg]) => ({
      value,
      label: cfg.label,
      hint: `${cfg.resource} [${cfg.changeType}]`
    })),
    required: true
  });
  if (p5.isCancel(selected)) {
    p5.cancel("Operation cancelled.");
    process.exit(0);
  }
  const clientState = await promptClientState();
  const results = [];
  let hasError = false;
  for (const value of selected) {
    const cfg = OUTLOOK_RESOURCES[value];
    const subSpin = p5.spinner();
    subSpin.start(`Creating subscription: ${cfg.label}...`);
    try {
      const subscription = await createGraphSubscription(
        accessToken,
        webhookUrl,
        cfg.resource,
        cfg.changeType,
        clientState,
        cfg.expiryMinutes
      );
      subSpin.stop(`Created: ${cfg.label}`);
      results.push(
        `${cfg.label}
  ID: ${subscription.id}
  Expires: ${subscription.expirationDateTime}`
      );
    } catch (error) {
      subSpin.stop(`Failed: ${cfg.label}`);
      p5.log.error(error instanceof Error ? error.message : String(error));
      hasError = true;
    }
  }
  if (results.length > 0) {
    p5.note(results.join("\n\n"), "Subscriptions created");
  }
  await saveWebhookSignature(accountKm, clientState);
  if (!hasError) {
    p5.log.success("Outlook webhook subscriptions set up successfully.");
  } else {
    p5.log.warn("Some subscriptions failed. Check errors above.");
  }
  p5.note(
    `ClientState: ${clientState}
Webhook URL: ${webhookUrl}

Note: Outlook subscriptions expire. Re-run this command to renew.`,
    "Setup complete"
  );
  p5.outro("Done!");
}
var OUTLOOK_RESOURCES;
var init_outlook = __esm({
  "src/lib/microsoft/outlook.ts"() {
    "use strict";
    init_load_config();
    init_prompts();
    init_credentials();
    init_graph();
    OUTLOOK_RESOURCES = {
      inboxMessages: {
        resource: "me/mailFolders/inbox/messages",
        changeType: "created",
        label: "Inbox Messages \u2014 new messages received",
        expiryMinutes: 4320
        // 3 days (max for mail)
      },
      sentMessages: {
        resource: "me/mailFolders/sentItems/messages",
        changeType: "created",
        label: "Sent Items \u2014 messages sent",
        expiryMinutes: 4320
      },
      newEvents: {
        resource: "me/events",
        changeType: "created",
        label: "Calendar Events \u2014 new events created",
        expiryMinutes: 4230
        // max for calendar
      },
      eventChanges: {
        resource: "me/events",
        changeType: "updated,deleted",
        label: "Calendar Events \u2014 updates & deletions",
        expiryMinutes: 4230
      },
      newContacts: {
        resource: "me/contacts",
        changeType: "created",
        label: "Contacts \u2014 new contacts created",
        expiryMinutes: 4230
        // max for contacts
      }
    };
  }
});

// src/lib/microsoft/sharepoint.ts
import * as p6 from "@clack/prompts";
function buildSiteUrl(siteId, subpath) {
  const isHostnamePath = siteId.includes(":") && !siteId.includes(",");
  return isHostnamePath ? `${GRAPH_API_BASE}/sites/${siteId}:/${subpath}` : `${GRAPH_API_BASE}/sites/${siteId}/${subpath}`;
}
async function fetchSiteLists(accessToken, siteId) {
  const response = await fetch(
    `${buildSiteUrl(siteId, "lists")}?$select=id,displayName,webUrl`,
    { headers: { Authorization: `Bearer ${accessToken}` } }
  );
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to fetch site lists: ${error}`);
  }
  const data = await response.json();
  return data.value ?? [];
}
async function createSharepointSubscription(accessToken, siteId, listId, notificationUrl, clientState, expiryDays) {
  const expirationDateTime = new Date(
    Date.now() + expiryDays * 24 * 60 * 60 * 1e3
  ).toISOString();
  const response = await fetch(
    buildSiteUrl(siteId, `lists/${encodeURIComponent(listId)}/subscriptions`),
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        notificationUrl,
        expirationDateTime,
        clientState
      })
    }
  );
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`SharePoint subscription failed: ${error}`);
  }
  return response.json();
}
async function runSharepointSubscribe({
  cwd
}) {
  const { internal, plugin: sharepointPlugin } = await loadInternalConfig(
    cwd,
    "Corsair \u2014 SharePoint Webhook Subscribe",
    "sharepoint",
    "SharePoint"
  );
  const tenantId = await promptTenantId();
  const { accessToken, accountKm } = await resolveAccessToken(
    "sharepoint",
    tenantId,
    internal
  );
  const webhookUrl = await promptWebhookUrl();
  const pluginOptions = sharepointPlugin.options;
  const defaultSiteId = pluginOptions?.siteId ?? "";
  const siteId = await p6.text({
    message: "Enter SharePoint site ID:",
    defaultValue: defaultSiteId,
    placeholder: "corsairdev.sharepoint.com:/sites/MySite  or  tenant.sharepoint.com,siteGuid,webGuid",
    validate: (v) => {
      if (!v || v.trim().length === 0) return "Site ID is required";
    }
  });
  if (p6.isCancel(siteId)) {
    p6.cancel("Operation cancelled.");
    process.exit(0);
  }
  const listSpin = p6.spinner();
  listSpin.start("Fetching lists from site...");
  let siteLists = [];
  try {
    siteLists = await fetchSiteLists(accessToken, siteId);
    listSpin.stop(
      `Found ${siteLists.length} list${siteLists.length === 1 ? "" : "s"}.`
    );
  } catch (error) {
    listSpin.stop("Could not fetch lists.");
    p6.log.warn(error instanceof Error ? error.message : String(error));
  }
  let listId;
  if (siteLists.length > 0) {
    const picked = await p6.select({
      message: "Select a list to subscribe to:",
      options: siteLists.map((l) => ({
        value: l.id,
        label: l.displayName,
        hint: l.id
      }))
    });
    if (p6.isCancel(picked)) {
      p6.cancel("Operation cancelled.");
      process.exit(0);
    }
    listId = picked;
  } else {
    const manualId = await p6.text({
      message: "Enter list ID (GUID) to subscribe to:",
      placeholder: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
      validate: (v) => {
        if (!v || v.trim().length === 0) return "List ID is required";
      }
    });
    if (p6.isCancel(manualId)) {
      p6.cancel("Operation cancelled.");
      process.exit(0);
    }
    listId = manualId;
  }
  const clientState = await promptClientState();
  const expiryDays = SHAREPOINT_MAX_EXPIRY_DAYS;
  const subSpin = p6.spinner();
  subSpin.start("Creating SharePoint webhook subscription...");
  let subscription;
  try {
    subscription = await createSharepointSubscription(
      accessToken,
      siteId,
      listId,
      webhookUrl,
      clientState,
      expiryDays
    );
  } catch (error) {
    subSpin.stop("Failed.");
    p6.log.error(error instanceof Error ? error.message : String(error));
    p6.outro("");
    process.exit(1);
  }
  subSpin.stop("Subscription created.");
  await saveWebhookSignature(accountKm, clientState);
  p6.note(
    `Subscription ID:  ${subscription.id}
Site ID:          ${siteId}
List ID:          ${listId}
Expires:          ${subscription.expirationDateTime}
ClientState:      ${clientState}
Webhook URL:      ${webhookUrl}

Note: SharePoint subscriptions expire after ${SHAREPOINT_MAX_EXPIRY_DAYS} days max. Re-run this command to renew.`,
    "SharePoint Subscription Created"
  );
  p6.outro("Done!");
}
var SHAREPOINT_MAX_EXPIRY_DAYS;
var init_sharepoint = __esm({
  "src/lib/microsoft/sharepoint.ts"() {
    "use strict";
    init_load_config();
    init_prompts();
    init_credentials();
    init_graph();
    SHAREPOINT_MAX_EXPIRY_DAYS = 180;
  }
});

// src/lib/microsoft/teams.ts
import * as p7 from "@clack/prompts";
async function fetchJoinedTeams(accessToken) {
  const response = await fetch(`${GRAPH_API_BASE}/me/joinedTeams`, {
    headers: { Authorization: `Bearer ${accessToken}` }
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to fetch teams: ${error}`);
  }
  const data = await response.json();
  return data.value ?? [];
}
async function fetchTeamChannels(accessToken, teamId) {
  const response = await fetch(
    `${GRAPH_API_BASE}/teams/${encodeURIComponent(teamId)}/channels`,
    { headers: { Authorization: `Bearer ${accessToken}` } }
  );
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to fetch channels: ${error}`);
  }
  const data = await response.json();
  return data.value ?? [];
}
async function fetchUserChats(accessToken) {
  const response = await fetch(`${GRAPH_API_BASE}/me/chats`, {
    headers: { Authorization: `Bearer ${accessToken}` }
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to fetch chats: ${error}`);
  }
  const data = await response.json();
  return data.value ?? [];
}
async function pickTeamId(accessToken, label) {
  const teamsSpin = p7.spinner();
  teamsSpin.start("Fetching your teams...");
  let teams = [];
  try {
    teams = await fetchJoinedTeams(accessToken);
    teamsSpin.stop(
      `Found ${teams.length} team${teams.length === 1 ? "" : "s"}.`
    );
  } catch (error) {
    teamsSpin.stop("Could not fetch teams.");
    p7.log.warn(error instanceof Error ? error.message : String(error));
  }
  if (teams.length > 0) {
    const picked = await p7.select({
      message: `[${label}] Select a team:`,
      options: teams.map((t) => ({
        value: t.id,
        label: t.displayName,
        hint: t.id
      }))
    });
    if (p7.isCancel(picked)) {
      p7.cancel("Operation cancelled.");
      process.exit(0);
    }
    return picked;
  }
  const manual = await p7.text({
    message: `[${label}] Enter team ID:`,
    placeholder: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
    validate: (v) => {
      if (!v || !v.trim()) return "Team ID is required";
    }
  });
  if (p7.isCancel(manual)) {
    p7.cancel("Operation cancelled.");
    process.exit(0);
  }
  return manual;
}
async function pickChannelId(accessToken, teamId, label) {
  const chanSpin = p7.spinner();
  chanSpin.start("Fetching channels...");
  let channels = [];
  try {
    channels = await fetchTeamChannels(accessToken, teamId);
    chanSpin.stop(
      `Found ${channels.length} channel${channels.length === 1 ? "" : "s"}.`
    );
  } catch (error) {
    chanSpin.stop("Could not fetch channels.");
    p7.log.warn(error instanceof Error ? error.message : String(error));
  }
  if (channels.length > 0) {
    const picked = await p7.select({
      message: `[${label}] Select a channel:`,
      options: channels.map((c) => ({
        value: c.id,
        label: c.displayName,
        hint: c.id
      }))
    });
    if (p7.isCancel(picked)) {
      p7.cancel("Operation cancelled.");
      process.exit(0);
    }
    return picked;
  }
  const manual = await p7.text({
    message: `[${label}] Enter channel ID:`,
    placeholder: "19:abc123@thread.skype",
    validate: (v) => {
      if (!v || !v.trim()) return "Channel ID is required";
    }
  });
  if (p7.isCancel(manual)) {
    p7.cancel("Operation cancelled.");
    process.exit(0);
  }
  return manual;
}
async function pickChatId(accessToken, label) {
  const chatSpin = p7.spinner();
  chatSpin.start("Fetching your chats...");
  let chats = [];
  try {
    chats = await fetchUserChats(accessToken);
    chatSpin.stop(
      `Found ${chats.length} chat${chats.length === 1 ? "" : "s"}.`
    );
  } catch (error) {
    chatSpin.stop("Could not fetch chats.");
    p7.log.warn(error instanceof Error ? error.message : String(error));
  }
  if (chats.length > 0) {
    const picked = await p7.select({
      message: `[${label}] Select a chat:`,
      options: chats.map((c) => ({
        value: c.id,
        label: c.topic ?? `(${c.chatType})`,
        hint: c.id
      }))
    });
    if (p7.isCancel(picked)) {
      p7.cancel("Operation cancelled.");
      process.exit(0);
    }
    return picked;
  }
  const manual = await p7.text({
    message: `[${label}] Enter chat ID:`,
    placeholder: "19:abc123@thread.v2",
    validate: (v) => {
      if (!v || !v.trim()) return "Chat ID is required";
    }
  });
  if (p7.isCancel(manual)) {
    p7.cancel("Operation cancelled.");
    process.exit(0);
  }
  return manual;
}
function buildTeamsResource(resourceType, ids) {
  switch (resourceType) {
    case "channelMessage":
      return `teams/${ids.teamId}/channels/${ids.channelId}/messages`;
    case "chatMessage":
      return `chats/${ids.chatId}/messages`;
    case "channelCreated":
      return `teams/${ids.teamId}/channels`;
    case "membershipChanged":
      return `teams/${ids.teamId}/members`;
  }
}
async function runTeamsSubscribe({
  cwd
}) {
  const { internal } = await loadInternalConfig(
    cwd,
    "Corsair \u2014 Microsoft Teams Webhook Subscribe",
    "teams",
    "Teams"
  );
  const tenantId = await promptTenantId();
  const { accessToken, accountKm } = await resolveAccessToken(
    "teams",
    tenantId,
    internal
  );
  const webhookUrl = await promptWebhookUrl();
  const selectedResourceTypes = await p7.multiselect({
    message: "Select resources to subscribe to:",
    options: [
      {
        value: "channelMessage",
        label: "Channel Messages  (teams/{id}/channels/{id}/messages)"
      },
      {
        value: "chatMessage",
        label: "Chat Messages     (chats/{id}/messages)"
      },
      {
        value: "channelCreated",
        label: "Channel Created   (teams/{id}/channels)"
      },
      {
        value: "membershipChanged",
        label: "Membership Changed (teams/{id}/members)"
      }
    ],
    required: true
  });
  if (p7.isCancel(selectedResourceTypes)) {
    p7.cancel("Operation cancelled.");
    process.exit(0);
  }
  const resourceConfigs = [];
  for (const resourceType of selectedResourceTypes) {
    const ids = {};
    if (resourceType === "chatMessage") {
      ids.chatId = await pickChatId(accessToken, resourceType);
    } else {
      ids.teamId = await pickTeamId(accessToken, resourceType);
      if (resourceType === "channelMessage") {
        ids.channelId = await pickChannelId(
          accessToken,
          ids.teamId,
          resourceType
        );
      }
    }
    resourceConfigs.push({ resourceType, ids });
  }
  const clientState = await promptClientState();
  const results = [];
  let hasError = false;
  for (const { resourceType, ids } of resourceConfigs) {
    const resource = buildTeamsResource(resourceType, ids);
    const expiryMinutes = TEAMS_MAX_EXPIRY_MINUTES[resourceType];
    const subSpin = p7.spinner();
    subSpin.start(`Creating subscription: ${resourceType}...`);
    try {
      const subscription = await createGraphSubscription(
        accessToken,
        webhookUrl,
        resource,
        "created,updated,deleted",
        clientState,
        expiryMinutes
      );
      subSpin.stop(`Created: ${resourceType}`);
      results.push(
        `${resourceType}
  Resource: ${resource}
  ID: ${subscription.id}
  Expires: ${subscription.expirationDateTime}`
      );
    } catch (error) {
      subSpin.stop(`Failed: ${resourceType}`);
      p7.log.error(error instanceof Error ? error.message : String(error));
      hasError = true;
    }
  }
  if (results.length > 0) {
    p7.note(results.join("\n\n"), "Subscriptions created");
  }
  await saveWebhookSignature(accountKm, clientState);
  if (!hasError) {
    p7.log.success("Teams webhook subscriptions set up successfully.");
  } else {
    p7.log.warn("Some subscriptions failed. Check errors above.");
  }
  p7.note(
    `ClientState: ${clientState}
Webhook URL: ${webhookUrl}

Note: Teams subscriptions expire. Re-run this command to renew.`,
    "Setup complete"
  );
  p7.outro("Done!");
}
var TEAMS_MAX_EXPIRY_MINUTES;
var init_teams = __esm({
  "src/lib/microsoft/teams.ts"() {
    "use strict";
    init_load_config();
    init_prompts();
    init_credentials();
    init_graph();
    TEAMS_MAX_EXPIRY_MINUTES = {
      channelMessage: 60,
      chatMessage: 60,
      channelCreated: 4230,
      membershipChanged: 60
    };
  }
});

// src/lib/microsoft/subscribe-microsoft.ts
var subscribe_microsoft_exports = {};
__export(subscribe_microsoft_exports, {
  runOnedriveSubscribe: () => runOnedriveSubscribe,
  runOutlookSubscribe: () => runOutlookSubscribe,
  runSharepointSubscribe: () => runSharepointSubscribe,
  runTeamsSubscribe: () => runTeamsSubscribe
});
var init_subscribe_microsoft = __esm({
  "src/lib/microsoft/subscribe-microsoft.ts"() {
    "use strict";
    init_onedrive();
    init_outlook();
    init_sharepoint();
    init_teams();
  }
});

// src/lib/google/calendar.ts
import * as crypto2 from "crypto";
import * as p8 from "@clack/prompts";
async function setupCalendarWatch(accessToken, webhookUrl, calendarId) {
  const watchSpin = p8.spinner();
  watchSpin.start("Starting Calendar watch...");
  const channelId = crypto2.randomUUID();
  const watchRes = await fetch(
    `${CALENDAR_API_BASE}/calendars/${encodeURIComponent(calendarId)}/events/watch`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        id: channelId,
        type: "web_hook",
        address: webhookUrl
      })
    }
  );
  if (!watchRes.ok) {
    const error = await watchRes.text();
    watchSpin.stop("Failed.");
    p8.log.error(`Calendar watch failed: ${error}`);
    p8.outro("");
    process.exit(1);
  }
  const data = await watchRes.json();
  watchSpin.stop("Watch started.");
  p8.note(
    `Channel ID: ${channelId}
Resource ID: ${data.resourceId}
Expiration: ${new Date(Number(data.expiration)).toISOString()}`,
    "Watch Response"
  );
}
var CALENDAR_API_BASE;
var init_calendar = __esm({
  "src/lib/google/calendar.ts"() {
    "use strict";
    CALENDAR_API_BASE = "https://www.googleapis.com/calendar/v3";
  }
});

// src/lib/google/drive.ts
import * as crypto3 from "crypto";
import * as p9 from "@clack/prompts";
async function setupDriveWatch(accessToken, webhookUrl) {
  const watchSpin = p9.spinner();
  watchSpin.start("Getting start page token...");
  const startPageTokenRes = await fetch(
    `${DRIVE_API_BASE}/changes/startPageToken`,
    {
      headers: { Authorization: `Bearer ${accessToken}` }
    }
  );
  if (!startPageTokenRes.ok) {
    const error = await startPageTokenRes.text();
    watchSpin.stop("Failed.");
    p9.log.error(`Failed to get start page token: ${error}`);
    p9.outro("");
    process.exit(1);
  }
  const startPageTokenData = await startPageTokenRes.json();
  watchSpin.message("Starting Drive watch...");
  const channelId = crypto3.randomUUID();
  const watchRes = await fetch(
    `${DRIVE_API_BASE}/changes/watch?pageToken=${startPageTokenData.startPageToken}`,
    {
      method: "POST",
      headers: {
        Authorization: `Bearer ${accessToken}`,
        "Content-Type": "application/json"
      },
      body: JSON.stringify({
        id: channelId,
        type: "web_hook",
        address: webhookUrl
      })
    }
  );
  if (!watchRes.ok) {
    const error = await watchRes.text();
    watchSpin.stop("Failed.");
    p9.log.error(`Drive watch failed: ${error}`);
    p9.outro("");
    process.exit(1);
  }
  const data = await watchRes.json();
  watchSpin.stop("Watch started.");
  p9.note(
    `Channel ID: ${channelId}
Resource ID: ${data.resourceId}
Start Page Token: ${startPageTokenData.startPageToken}
Expiration: ${new Date(Number(data.expiration)).toISOString()}`,
    "Watch Response"
  );
}
var DRIVE_API_BASE;
var init_drive = __esm({
  "src/lib/google/drive.ts"() {
    "use strict";
    DRIVE_API_BASE = "https://www.googleapis.com/drive/v3";
  }
});

// src/lib/google/gmail.ts
import * as p10 from "@clack/prompts";
async function setupGmailWatch(accessToken, topicName) {
  const watchSpin = p10.spinner();
  watchSpin.start("Starting Gmail watch...");
  const response = await fetch(`${GMAIL_API_BASE}/users/me/watch`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json"
    },
    body: JSON.stringify({
      topicName,
      labelIds: ["INBOX"]
    })
  });
  if (!response.ok) {
    const error = await response.text();
    watchSpin.stop("Failed.");
    p10.log.error(`Gmail watch failed: ${error}`);
    p10.outro("");
    process.exit(1);
  }
  const data = await response.json();
  watchSpin.stop("Watch started.");
  p10.note(
    `History ID: ${data.historyId}
Expiration: ${new Date(Number(data.expiration)).toISOString()}`,
    "Watch Response"
  );
}
var GMAIL_API_BASE;
var init_gmail = __esm({
  "src/lib/google/gmail.ts"() {
    "use strict";
    GMAIL_API_BASE = "https://gmail.googleapis.com/gmail/v1";
  }
});

// src/lib/google/shared.ts
import { execSync } from "child_process";
async function refreshGoogleAccessToken(clientId, clientSecret, refreshToken) {
  const response = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      client_id: clientId,
      client_secret: clientSecret,
      refresh_token: refreshToken,
      grant_type: "refresh_token"
    })
  });
  if (!response.ok) {
    const error = await response.text();
    throw new Error(`Failed to refresh access token: ${error}`);
  }
  const data = await response.json();
  return data.access_token;
}
function copyToClipboard(text6) {
  try {
    const platform = process.platform;
    if (platform === "darwin") {
      execSync("pbcopy", { input: text6 });
      return true;
    } else if (platform === "linux") {
      try {
        execSync("xclip -selection clipboard", { input: text6 });
        return true;
      } catch {
        try {
          execSync("xsel --clipboard --input", { input: text6 });
          return true;
        } catch {
          return false;
        }
      }
    } else if (platform === "win32") {
      execSync("clip", { input: text6 });
      return true;
    }
    return false;
  } catch {
    return false;
  }
}
var init_shared = __esm({
  "src/lib/google/shared.ts"() {
    "use strict";
  }
});

// src/lib/google/sheets.ts
import * as p11 from "@clack/prompts";
async function setupSheetsWatch(webhookUrl) {
  const appsScriptCode = APPS_SCRIPT_TEMPLATE(webhookUrl);
  p11.log.info("Google Sheets does not have a native watch API.");
  p11.log.info(
    "Instead, you need to set up Google Apps Script triggers in your spreadsheet."
  );
  p11.log.info("");
  const shouldCopy = await p11.confirm({
    message: "Copy Apps Script code to clipboard?",
    initialValue: true
  });
  let copied = false;
  if (shouldCopy) {
    copied = copyToClipboard(appsScriptCode);
    if (copied) {
      p11.log.success("Apps Script code copied to clipboard!");
    } else {
      p11.log.warn("Failed to copy to clipboard. Code will be displayed below.");
    }
  }
  p11.log.info("");
  p11.log.info("Setup Instructions:");
  p11.log.info("1. Open your Google Sheet");
  p11.log.info("2. Go to Extensions > Apps Script");
  p11.log.info("3. Paste the code into the script editor");
  p11.log.info(
    '4. Run the "createTriggers" function once to install the triggers'
  );
  p11.log.info("5. Authorize the script when prompted");
  p11.log.info("");
  if (!copied) {
    p11.log.info("Apps Script Code:");
    p11.log.info("");
    p11.log.info(appsScriptCode);
    p11.log.info("");
  }
  p11.note(
    `Webhook URL: ${webhookUrl}

The Apps Script code ${copied ? "has been copied to your clipboard" : "is displayed above"}. Paste it into your Google Sheet's Apps Script editor.`,
    "Google Sheets Setup"
  );
}
var APPS_SCRIPT_TEMPLATE;
var init_sheets = __esm({
  "src/lib/google/sheets.ts"() {
    "use strict";
    init_shared();
    APPS_SCRIPT_TEMPLATE = (webhookUrl) => `var WEBHOOK_URL = "${webhookUrl}";

function onEditTrigger(e) {
  if (!e || !e.range) return;

  var sheet = e.range.getSheet();
  var payload = {
    spreadsheetId: e.source.getId(),
    sheetName: sheet.getName(),
    range: e.range.getA1Notation(),
    values: e.range.getValues(),
    eventType: "rangeUpdated",
    timestamp: new Date().toISOString()
  };

  UrlFetchApp.fetch(WEBHOOK_URL, {
    method: "post",
    contentType: "application/json",
    payload: JSON.stringify(payload)
  });
}

function onChangeTrigger(e) {
  if (!e) return;

  var ss = SpreadsheetApp.getActiveSpreadsheet();
  var sheet = ss.getActiveSheet();
  var lastRow = sheet.getLastRow();
  var lastCol = sheet.getLastColumn();
  if (lastRow === 0 || lastCol === 0) return;

  var values = sheet.getRange(lastRow, 1, 1, lastCol).getValues();

  var payload = {
    spreadsheetId: ss.getId(),
    sheetName: sheet.getName(),
    range: "A" + lastRow + ":" + String.fromCharCode(64 + lastCol) + lastRow,
    values: values,
    eventType: "rangeUpdated",
    timestamp: new Date().toISOString()
  };

  UrlFetchApp.fetch(WEBHOOK_URL, {
    method: "post",
    contentType: "application/json",
    payload: JSON.stringify(payload)
  });
}

function createTriggers() {
  var ss = SpreadsheetApp.getActiveSpreadsheet();

  ScriptApp.newTrigger("onEditTrigger").forSpreadsheet(ss).onEdit().create();
  ScriptApp.newTrigger("onChangeTrigger").forSpreadsheet(ss).onChange().create();

  Logger.log("Triggers created successfully!");
}

function removeTriggers() {
  var triggers = ScriptApp.getProjectTriggers();
  triggers.forEach(function(trigger) {
    ScriptApp.deleteTrigger(trigger);
  });
  Logger.log("All triggers removed.");
}`;
  }
});

// src/lib/google/index.ts
import * as p12 from "@clack/prompts";
import {
  CORSAIR_INTERNAL as CORSAIR_INTERNAL2,
  createAccountKeyManager as createAccountKeyManager2,
  createIntegrationKeyManager as createIntegrationKeyManager2
} from "corsair/core";
async function extractInternalConfig2(cwd) {
  const instance = await getCorsairInstance({ cwd, shouldThrowOnError: true });
  const internal = instance[CORSAIR_INTERNAL2];
  if (!internal) {
    throw new Error(
      "Could not read internal config from Corsair instance. Make sure you are using the latest version of corsair."
    );
  }
  return internal;
}
async function runGoogleSubscribe({
  cwd,
  pluginId: preselectedPluginId
}) {
  p12.intro("Corsair \u2014 Google Webhook Subscribe");
  const spin = p12.spinner();
  spin.start("Loading corsair instance...");
  let internal;
  try {
    internal = await extractInternalConfig2(cwd);
  } catch (error) {
    spin.stop("Failed to load.");
    p12.log.error(error instanceof Error ? error.message : String(error));
    p12.outro("");
    process.exit(1);
  }
  const { plugins, database, kek } = internal;
  if (!database) {
    spin.stop("Failed.");
    p12.log.error("No database adapter configured.");
    p12.outro("");
    process.exit(1);
  }
  const googlePlugins = plugins.filter((pl) => {
    const opts = pl.options;
    return opts?.authType === "oauth_2" && GOOGLE_PLUGINS.includes(pl.id);
  });
  if (googlePlugins.length === 0) {
    spin.stop("No Google OAuth2 plugins found.");
    p12.outro("");
    process.exit(1);
  }
  spin.stop(
    `Loaded. Found ${googlePlugins.length} Google plugin${googlePlugins.length === 1 ? "" : "s"}.`
  );
  let pluginType;
  if (preselectedPluginId && GOOGLE_PLUGINS.includes(preselectedPluginId)) {
    pluginType = preselectedPluginId;
  } else {
    const selected = await p12.select({
      message: "Select a Google integration:",
      options: googlePlugins.map((pl) => ({ value: pl.id, label: pl.id }))
    });
    if (p12.isCancel(selected)) {
      p12.cancel("Operation cancelled.");
      process.exit(0);
    }
    pluginType = selected;
  }
  if (pluginType === "googlesheets") {
    const webhookUrl = await p12.text({
      message: "Enter webhook URL:",
      placeholder: "https://example.com/api/webhook",
      validate: (v) => {
        if (!v || v.trim().length === 0) return "Webhook URL is required";
        if (!v.startsWith("http://") && !v.startsWith("https://"))
          return "Webhook URL must start with http:// or https://";
      }
    });
    if (p12.isCancel(webhookUrl)) {
      p12.cancel("Operation cancelled.");
      process.exit(0);
    }
    await setupSheetsWatch(webhookUrl);
    p12.outro("Done!");
    return;
  }
  const tenantId = await promptTenantId();
  const credSpin = p12.spinner();
  credSpin.start("Fetching credentials...");
  try {
    const integrationKm = createIntegrationKeyManager2({
      authType: "oauth_2",
      integrationName: pluginType,
      kek,
      database
    });
    const accountKm = createAccountKeyManager2({
      authType: "oauth_2",
      integrationName: pluginType,
      tenantId,
      kek,
      database
    });
    const clientId = await integrationKm.get_client_id();
    const clientSecret = await integrationKm.get_client_secret();
    const refreshToken = await accountKm.get_refresh_token();
    if (!clientId || !clientSecret || !refreshToken) {
      credSpin.stop("Missing credentials.");
      p12.log.error(
        'Client ID, Client Secret, and Refresh Token are required. Use "corsair auth" to set them.'
      );
      p12.outro("");
      process.exit(1);
    }
    const accessToken = await refreshGoogleAccessToken(
      clientId,
      clientSecret,
      refreshToken
    );
    credSpin.stop("Credentials loaded.");
    if (pluginType === "gmail") {
      const topicName = await p12.text({
        message: "Enter Pub/Sub topic name:",
        placeholder: "projects/my-project/topics/my-topic",
        validate: (v) => {
          if (!v || v.trim().length === 0) return "Topic name is required";
        }
      });
      if (p12.isCancel(topicName)) {
        p12.cancel("Operation cancelled.");
        process.exit(0);
      }
      await setupGmailWatch(accessToken, topicName);
    } else if (pluginType === "googledrive") {
      const webhookUrl = await p12.text({
        message: "Enter webhook URL:",
        placeholder: "https://example.com/api/webhook",
        validate: (v) => {
          if (!v || v.trim().length === 0) return "Webhook URL is required";
        }
      });
      if (p12.isCancel(webhookUrl)) {
        p12.cancel("Operation cancelled.");
        process.exit(0);
      }
      await setupDriveWatch(accessToken, webhookUrl);
    } else if (pluginType === "googlecalendar") {
      const webhookUrl = await p12.text({
        message: "Enter webhook URL:",
        placeholder: "https://example.com/api/webhook",
        validate: (v) => {
          if (!v || v.trim().length === 0) return "Webhook URL is required";
        }
      });
      if (p12.isCancel(webhookUrl)) {
        p12.cancel("Operation cancelled.");
        process.exit(0);
      }
      const calendarId = await p12.text({
        message: "Enter calendar ID:",
        defaultValue: "primary",
        placeholder: "primary"
      });
      if (p12.isCancel(calendarId)) {
        p12.cancel("Operation cancelled.");
        process.exit(0);
      }
      await setupCalendarWatch(
        accessToken,
        webhookUrl,
        calendarId
      );
    } else {
      p12.log.error(`Unsupported Google plugin: ${pluginType}`);
      process.exit(1);
    }
  } catch (error) {
    credSpin.stop("Failed.");
    p12.log.error(
      `Error: ${error instanceof Error ? error.message : String(error)}`
    );
    process.exit(1);
  }
  p12.outro("Done!");
}
var GOOGLE_PLUGINS;
var init_google = __esm({
  "src/lib/google/index.ts"() {
    "use strict";
    init_corsair_instance();
    init_prompts();
    init_calendar();
    init_drive();
    init_gmail();
    init_shared();
    init_sheets();
    GOOGLE_PLUGINS = [
      "gmail",
      "googledrive",
      "googlecalendar",
      "googlesheets"
    ];
  }
});

// src/lib/google/subscribe-google.ts
var subscribe_google_exports = {};
__export(subscribe_google_exports, {
  runGoogleSubscribe: () => runGoogleSubscribe
});
var init_subscribe_google = __esm({
  "src/lib/google/subscribe-google.ts"() {
    "use strict";
    init_google();
  }
});

// src/auth.ts
var auth_exports = {};
__export(auth_exports, {
  runAuth: () => runAuth
});
import { spawn } from "child_process";
import * as crypto4 from "crypto";
import * as fs3 from "fs";
import * as http from "http";
import * as https from "https";
import * as net from "net";
import * as os from "os";
import * as path3 from "path";
import * as querystring from "querystring";
import {
  CORSAIR_INTERNAL as CORSAIR_INTERNAL3,
  createAccountKeyManager as createAccountKeyManager3,
  createIntegrationKeyManager as createIntegrationKeyManager3,
  encryptDEK,
  exchangeCodeForTokens,
  generateDEK
} from "corsair/core";
import { createCorsairOrm } from "corsair/orm";
function out(data) {
  console.log(JSON.stringify(data));
}
function authStateFile(sessionId) {
  return path3.join(os.tmpdir(), `corsair-auth-${sessionId}.json`);
}
function writeAuthState(sessionId, state) {
  fs3.writeFileSync(authStateFile(sessionId), JSON.stringify(state));
}
function pollAuthState(sessionId, timeoutMs) {
  return new Promise((resolve, reject) => {
    const file = authStateFile(sessionId);
    const start = Date.now();
    const interval = setInterval(() => {
      try {
        const state = JSON.parse(fs3.readFileSync(file, "utf8"));
        clearInterval(interval);
        resolve(state);
      } catch {
        if (Date.now() - start > timeoutMs) {
          clearInterval(interval);
          reject(new Error("Timed out waiting for OAuth server to start"));
        }
      }
    }, 100);
  });
}
function getOAuthConfigForPlugin(plugin) {
  return plugin.oauthConfig ?? null;
}
function getAuthType(plugin) {
  return plugin.options?.authType;
}
function getCustomIntegrationFields(plugin, authType) {
  const authConfig = plugin.authConfig;
  return authConfig?.[authType]?.integration ?? [];
}
function getCustomAccountFields(plugin, authType) {
  const authConfig = plugin.authConfig;
  return authConfig?.[authType]?.account ?? [];
}
function findFreePort() {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      const port = typeof address === "object" && address ? address.port : 0;
      server.close((err) => {
        if (err) reject(err);
        else resolve(port);
      });
    });
    server.on("error", reject);
  });
}
function waitForOAuthCode(port) {
  return new Promise((resolve, reject) => {
    const server = http.createServer((req, res) => {
      const url = new URL(req.url ?? "/", `http://localhost:${port}`);
      const code = url.searchParams.get("code");
      const error = url.searchParams.get("error");
      if (error) {
        res.writeHead(400, { "Content-Type": "text/html" });
        res.end(
          `<html><body><h2>Authorization failed: ${error}</h2><p>You can close this tab.</p></body></html>`
        );
        server.close();
        reject(new Error(`OAuth error: ${error}`));
        return;
      }
      if (code) {
        res.writeHead(200, { "Content-Type": "text/html" });
        res.end(
          "<html><body><h2>Authorization successful!</h2><p>You can close this tab and return to the terminal.</p></body></html>"
        );
        server.close();
        resolve(code);
      } else {
        res.writeHead(400, { "Content-Type": "text/html" });
        res.end(
          "<html><body><h2>No code received.</h2><p>You can close this tab.</p></body></html>"
        );
      }
    });
    server.listen(port, "127.0.0.1", () => {
    });
    server.on("error", reject);
  });
}
async function extractInternalConfig3(cwd) {
  const instance = await getCorsairInstance({ cwd, shouldThrowOnError: true });
  const internal = instance[CORSAIR_INTERNAL3];
  if (!internal) {
    throw new Error(
      "Could not read internal config from Corsair instance. Make sure you are using the latest version of corsair."
    );
  }
  return internal;
}
async function ensureIntegration(database, pluginId, kek) {
  const orm = createCorsairOrm(database);
  const existing = await orm.integrations.findByName(pluginId);
  if (existing) return { id: existing.id };
  const dek = generateDEK();
  const encryptedDek = await encryptDEK(dek, kek);
  const row = await orm.integrations.create({
    name: pluginId,
    config: {},
    dek: encryptedDek
  });
  return { id: row.id };
}
async function ensureAccount(database, integrationId, tenantId, kek) {
  const orm = createCorsairOrm(database);
  const existing = await orm.accounts.findOne({
    tenant_id: tenantId,
    integration_id: integrationId
  });
  if (existing) return;
  const dek = generateDEK();
  const encryptedDek = await encryptDEK(dek, kek);
  await orm.accounts.create({
    tenant_id: tenantId,
    integration_id: integrationId,
    config: {},
    dek: encryptedDek
  });
}
async function oauthExchangeCode(database, plugin, kek, tenantId, code, redirectUri, clientId, clientSecret, oauthCfg) {
  let tokens;
  try {
    tokens = await exchangeCodeForTokens(
      code,
      clientId,
      clientSecret,
      oauthCfg,
      redirectUri
    );
  } catch (err) {
    out({
      error: `Token exchange failed: ${err instanceof Error ? err.message : String(err)}`
    });
    return false;
  }
  if (!tokens.access_token) {
    out({
      error: `No access_token in response from ${oauthCfg.providerName}.`
    });
    return false;
  }
  const extraAccountFields = getCustomAccountFields(plugin, "oauth_2");
  const accountKm = createAccountKeyManager3({
    authType: "oauth_2",
    integrationName: plugin.id,
    tenantId,
    kek,
    database,
    extraAccountFields
  });
  await accountKm.set_access_token(tokens.access_token);
  if (tokens.refresh_token)
    await accountKm.set_refresh_token(tokens.refresh_token);
  if (tokens.expires_in)
    await accountKm.set_expires_at(
      (Math.floor(Date.now() / 1e3) + tokens.expires_in).toString()
    );
  out({ status: "success", plugin: plugin.id, tenant: tenantId });
  return true;
}
async function oauthWithCode(database, plugin, kek, tenantId, code) {
  const oauthCfg = getOAuthConfigForPlugin(plugin);
  if (!oauthCfg) {
    out({ error: `No oauthConfig defined on plugin '${plugin.id}'.` });
    return;
  }
  const extraFields = getCustomIntegrationFields(plugin, "oauth_2");
  const integrationKm = createIntegrationKeyManager3({
    authType: "oauth_2",
    integrationName: plugin.id,
    kek,
    database,
    extraIntegrationFields: extraFields
  });
  const clientId = await integrationKm.get_client_id();
  const clientSecret = await integrationKm.get_client_secret();
  const redirectUri = await integrationKm.get_redirect_url() ?? "";
  if (!clientId || !clientSecret) {
    out({
      error: `client_id and client_secret must be set before exchanging a code.`
    });
    return;
  }
  await oauthExchangeCode(
    database,
    plugin,
    kek,
    tenantId,
    code,
    redirectUri,
    clientId,
    clientSecret,
    oauthCfg
  );
}
async function oauthListen(database, plugin, kek, tenantId, sessionId) {
  const oauthCfg = getOAuthConfigForPlugin(plugin);
  if (!oauthCfg) {
    writeAuthState(sessionId, {
      status: "error",
      error: `No oauthConfig defined on plugin '${plugin.id}'.`
    });
    return;
  }
  const extraFields = getCustomIntegrationFields(plugin, "oauth_2");
  const integrationKm = createIntegrationKeyManager3({
    authType: "oauth_2",
    integrationName: plugin.id,
    kek,
    database,
    extraIntegrationFields: extraFields
  });
  const clientId = await integrationKm.get_client_id();
  if (!clientId) {
    writeAuthState(sessionId, {
      status: "error",
      error: `client_id not set for '${plugin.id}'.`
    });
    return;
  }
  const clientSecret = await integrationKm.get_client_secret();
  if (!clientSecret) {
    writeAuthState(sessionId, {
      status: "error",
      error: `client_secret not set for '${plugin.id}'.`
    });
    return;
  }
  let redirectUri;
  let localhostPort;
  if (oauthCfg.requiresRegisteredRedirect) {
    const stored = await integrationKm.get_redirect_url();
    if (!stored) {
      writeAuthState(sessionId, {
        status: "error",
        error: `redirect_url required for '${plugin.id}'.`
      });
      return;
    }
    const match = stored.match(/^https?:\/\/(?:localhost|127\.0\.0\.1):(\d+)/);
    if (!match) {
      const authParams2 = {
        client_id: clientId,
        redirect_uri: stored,
        response_type: "code",
        scope: oauthCfg.scopes.join(" "),
        ...oauthCfg.authParams
      };
      const authUrl2 = `${oauthCfg.authUrl}?${querystring.stringify(authParams2)}`;
      writeAuthState(sessionId, {
        status: "needs_code",
        url: authUrl2,
        redirectUri: stored
      });
      return;
    }
    redirectUri = stored;
    localhostPort = parseInt(match[1], 10);
  } else {
    localhostPort = await findFreePort();
    redirectUri = `http://localhost:${localhostPort}`;
  }
  const authParams = {
    client_id: clientId,
    redirect_uri: redirectUri,
    response_type: "code",
    scope: oauthCfg.scopes.join(" "),
    ...oauthCfg.authParams
  };
  const authUrl = `${oauthCfg.authUrl}?${querystring.stringify(authParams)}`;
  writeAuthState(sessionId, { status: "listening", url: authUrl });
  const timeout = setTimeout(
    () => {
      writeAuthState(sessionId, {
        status: "error",
        error: "OAuth session timed out after 10 minutes."
      });
      process.exit(0);
    },
    10 * 60 * 1e3
  );
  timeout.unref();
  let code;
  try {
    code = await waitForOAuthCode(localhostPort);
  } catch (err) {
    writeAuthState(sessionId, {
      status: "error",
      error: `Authorization failed: ${err instanceof Error ? err.message : String(err)}`
    });
    return;
  }
  const success = await oauthExchangeCode(
    database,
    plugin,
    kek,
    tenantId,
    code,
    redirectUri,
    clientId,
    clientSecret,
    oauthCfg
  );
  writeAuthState(
    sessionId,
    success ? { status: "complete" } : { status: "error", error: "Token exchange failed." }
  );
}
async function oauthCollect(database, plugin, kek, tenantId, sessionId) {
  const extraAccountFields = getCustomAccountFields(plugin, "oauth_2");
  const accountKm = createAccountKeyManager3({
    authType: "oauth_2",
    integrationName: plugin.id,
    tenantId,
    kek,
    database,
    extraAccountFields
  });
  const accessToken = await accountKm.get_access_token();
  if (accessToken) {
    try {
      fs3.unlinkSync(authStateFile(sessionId));
    } catch {
    }
    out({ status: "success", plugin: plugin.id, tenant: tenantId });
    return;
  }
  try {
    const state = JSON.parse(
      fs3.readFileSync(authStateFile(sessionId), "utf8")
    );
    if (state.status === "error") {
      out({ status: "error", error: state.error });
      return;
    }
  } catch {
  }
  out({
    status: "pending",
    message: "User has not completed sign-in yet. Try again after the user visits the auth URL."
  });
}
async function getCredentials(database, plugin, kek, tenantId) {
  const authType = getAuthType(plugin);
  const result = {};
  if (authType === "oauth_2") {
    const extraIntegrationFields = getCustomIntegrationFields(plugin, authType);
    const integrationKm = createIntegrationKeyManager3({
      authType,
      integrationName: plugin.id,
      kek,
      database,
      extraIntegrationFields
    });
    result.client_id = await integrationKm.get_client_id();
    result.client_secret = await integrationKm.get_client_secret();
    result.redirect_url = await integrationKm.get_redirect_url();
    for (const field of extraIntegrationFields) {
      const fn = integrationKm[`get_${field}`];
      if (typeof fn === "function")
        result[field] = await fn();
    }
    const extraAccountFields = getCustomAccountFields(plugin, authType);
    const accountKm = createAccountKeyManager3({
      authType,
      integrationName: plugin.id,
      tenantId,
      kek,
      database,
      extraAccountFields
    });
    result.access_token = await accountKm.get_access_token();
    result.refresh_token = await accountKm.get_refresh_token();
    result.expires_at = await accountKm.get_expires_at();
    result.webhook_signature = await accountKm.get_webhook_signature();
    for (const field of extraAccountFields) {
      const fn = accountKm[`get_${field}`];
      if (typeof fn === "function")
        result[field] = await fn();
    }
  } else if (authType === "api_key") {
    const extraAccountFields = getCustomAccountFields(plugin, authType);
    const accountKm = createAccountKeyManager3({
      authType,
      integrationName: plugin.id,
      tenantId,
      kek,
      database,
      extraAccountFields
    });
    result.api_key = await accountKm.get_api_key();
    result.webhook_signature = await accountKm.get_webhook_signature();
  } else if (authType === "bot_token") {
    const extraAccountFields = getCustomAccountFields(plugin, authType);
    const accountKm = createAccountKeyManager3({
      authType,
      integrationName: plugin.id,
      tenantId,
      kek,
      database,
      extraAccountFields
    });
    result.bot_token = await accountKm.get_bot_token();
    result.webhook_signature = await accountKm.get_webhook_signature();
  }
  const masked = {};
  for (const [k, v] of Object.entries(result)) {
    if (!v) {
      masked[k] = null;
      continue;
    }
    masked[k] = v.length <= 9 ? "***" : `${v.slice(0, 6)}...${v.slice(-3)}`;
  }
  out({ plugin: plugin.id, tenant: tenantId, credentials: masked });
}
async function runAuth({
  cwd,
  pluginId: pluginIdArg,
  tenantId: tenantIdArg,
  code: codeArg,
  credentials: showCredentials = false,
  listen: listenMode = false,
  collect: collectMode = false,
  sessionId
}) {
  let internal;
  try {
    internal = await extractInternalConfig3(cwd);
  } catch (error) {
    out({ error: error instanceof Error ? error.message : String(error) });
    process.exit(1);
  }
  const { plugins, database, kek } = internal;
  if (!database) {
    out({ error: "No database adapter configured." });
    process.exit(1);
  }
  if (!pluginIdArg) {
    out({
      error: "No plugin specified. Use --plugin=<id>.",
      available: plugins.map((pl) => ({
        id: pl.id,
        authType: getAuthType(pl)
      }))
    });
    process.exit(1);
  }
  const plugin = plugins.find((pl) => pl.id === pluginIdArg);
  if (!plugin) {
    out({
      error: `Plugin '${pluginIdArg}' not found.`,
      available: plugins.map((pl) => pl.id)
    });
    process.exit(1);
  }
  const integration = await ensureIntegration(database, plugin.id, kek);
  const tenantId = tenantIdArg ?? "default";
  await ensureAccount(database, integration.id, tenantId, kek);
  if (showCredentials) {
    await getCredentials(database, plugin, kek, tenantId);
    return;
  }
  const authType = getAuthType(plugin);
  if (authType !== "oauth_2") {
    out({
      error: `'corsair auth' is for OAuth flows. Plugin '${plugin.id}' uses '${authType}'. Set credentials via: corsair setup --${plugin.id} <field>=<value>`
    });
    return;
  }
  if (listenMode) {
    if (!sessionId) {
      out({ error: "--listen requires --session=<id>." });
      process.exit(1);
    }
    await oauthListen(database, plugin, kek, tenantId, sessionId);
    return;
  }
  if (collectMode) {
    if (!sessionId) {
      out({ error: "--collect requires --session=<id>." });
      process.exit(1);
    }
    await oauthCollect(database, plugin, kek, tenantId, sessionId);
    return;
  }
  if (codeArg) {
    await oauthWithCode(database, plugin, kek, tenantId, codeArg);
    return;
  }
  const session = crypto4.randomUUID().slice(0, 8);
  const needsQuoting = tenantIdArg && /[^a-zA-Z0-9_\-.]/.test(tenantIdArg);
  const tenantFlag = tenantIdArg && tenantIdArg !== "default" ? `--tenant=${needsQuoting ? `"${tenantIdArg}"` : tenantIdArg}` : "";
  const spawnOpts = {
    detached: true,
    stdio: "ignore",
    cwd: process.cwd(),
    env: process.env
  };
  const child = spawn(
    process.execPath,
    [
      ...process.execArgv,
      process.argv[1] ?? "",
      "auth",
      `--plugin=${plugin.id}`,
      "--listen",
      `--session=${session}`,
      ...tenantFlag ? [tenantFlag] : []
    ],
    spawnOpts
  );
  child.unref();
  let state;
  try {
    state = await pollAuthState(session, 5e3);
  } catch (err) {
    out({
      error: `Failed to start OAuth server: ${err instanceof Error ? err.message : String(err)}`
    });
    return;
  }
  if (state.status === "needs_code") {
    console.log(`
Authorize ${plugin.id}
`);
    console.log(`  Open this URL in your browser:
`);
    console.log(`    ${state.url}
`);
    console.log(
      `  Once you've signed in, copy the code from the redirect URL and run:
`
    );
    console.log(`    pnpm corsair auth --plugin=${plugin.id} --code=CODE
`);
    return;
  }
  if (state.status === "error") {
    out({ error: state.error });
    return;
  }
  const collectCmd = [
    `pnpm corsair auth --plugin=${plugin.id}`,
    "--collect",
    `--session=${session}`,
    ...tenantFlag ? [tenantFlag] : []
  ].join(" ");
  console.log(`
Authorize ${plugin.id}
`);
  console.log(`  Open this URL in your browser:
`);
  console.log(`    ${state.url}
`);
  console.log(`  Once you've signed in, run:
`);
  console.log(`    ${collectCmd}
`);
}
var init_auth = __esm({
  "src/auth.ts"() {
    "use strict";
    init_corsair_instance();
  }
});

// src/index.ts
import { realpathSync } from "fs";
import { fileURLToPath } from "url";
import { Command } from "commander";

// package.json
var version = "0.1.19";

// src/utils/subscription.ts
async function runWebhookSubscription(cwd, pluginId) {
  const microsoftPluginRunner = {
    outlook: "runOutlookSubscribe",
    sharepoint: "runSharepointSubscribe",
    teams: "runTeamsSubscribe",
    onedrive: "runOnedriveSubscribe"
  };
  if (pluginId in microsoftPluginRunner) {
    const mod = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
    const fn = mod[microsoftPluginRunner[pluginId]];
    if (typeof fn === "function") {
      await fn({ cwd });
      return;
    }
  }
  if (["gmail", "googledrive", "googlecalendar", "googlesheets"].includes(
    pluginId
  )) {
    const { runGoogleSubscribe: runGoogleSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_google(), subscribe_google_exports));
    await runGoogleSubscribe2({ cwd, pluginId });
    return;
  }
  console.error(
    `[#corsair]: Webhook subscription not supported for plugin '${pluginId}'.`
  );
  process.exit(1);
}

// src/commands/base.command.ts
import { Option } from "commander";
var BaseCommand = class {
  action({}) {
  }
  getAliases() {
    return [];
  }
  getArguments() {
    return [];
  }
  getOptions() {
    return [];
  }
  getUsage() {
    return "";
  }
  getSubCommands() {
    return [];
  }
  /** When true, unknown flags (e.g. `--slack`) are passed through to the action. */
  allowUnknownOptions() {
    return false;
  }
  prepare = (program) => {
    const command = program.command(this.getName()).description(this.getDescription()).allowUnknownOption(this.allowUnknownOptions());
    const usage = this.getUsage();
    if (usage) {
      command.usage(usage);
    }
    this.getAliases().forEach((alias) => command.alias(alias));
    this.getOptions().forEach((option) => {
      const newOption = new Option(
        `${option.short}, ${option.long}`,
        option.description
      ).default(option.defaultValue);
      option.choices && option.choices.length > 0 && newOption.choices(option.choices);
      command.addOption(newOption);
    });
    this.getArguments().forEach(
      (argument) => command.argument(argument.name, argument.description)
    );
    command.action(async (...data) => {
      const argsCount = this.getArguments().length;
      const commandOptions = data[argsCount];
      const args = data.slice(0, argsCount);
      await this.action({ args, options: commandOptions });
    });
    this.getSubCommands().forEach((subCommand) => {
      subCommand.prepare(command);
    });
  };
};

// src/commands/auth.command.ts
var AuthCommand = class extends BaseCommand {
  getName() {
    return "auth";
  }
  getDescription() {
    return "Run plugin auth flows";
  }
  getOptions() {
    return [
      { short: "-p", long: "--plugin <id>", description: "Plugin id" },
      { short: "-t", long: "--tenant <id>", description: "Tenant id for auth" },
      {
        short: "-c",
        long: "--code <code>",
        description: "OAuth code to exchange"
      },
      {
        short: "-s",
        long: "--session <id>",
        description: "Session id for local auth state"
      },
      {
        short: "-C",
        long: "--credentials",
        description: "Show credential status"
      },
      {
        short: "-w",
        long: "--webhook",
        description: "Run webhook subscription setup"
      },
      {
        short: "-l",
        long: "--listen",
        description: "Listen for callback only"
      },
      {
        short: "-x",
        long: "--collect",
        description: "Collect auth response from listener"
      }
    ];
  }
  async action({ options }) {
    const cwd = process.cwd();
    if (options.webhook) {
      if (!options.plugin) {
        console.error("[#corsair]: --webhook requires --plugin=<id>.");
        process.exit(1);
      }
      await runWebhookSubscription(cwd, options.plugin);
      return;
    }
    const { runAuth: runAuth2 } = await Promise.resolve().then(() => (init_auth(), auth_exports));
    await runAuth2({
      cwd,
      pluginId: options.plugin,
      tenantId: options.tenant,
      code: options.code,
      credentials: options.credentials,
      listen: options.listen,
      collect: options.collect,
      sessionId: options.session
    });
  }
};

// src/commands/list.command.ts
init_corsair_instance();
import { listOperations } from "corsair";
var ListCommand = class extends BaseCommand {
  getName() {
    return "list";
  }
  getDescription() {
    return "List available operations";
  }
  getOptions() {
    return [
      {
        short: "-p",
        long: "--plugin <id>",
        description: "Filter by plugin id"
      },
      {
        short: "-t",
        long: "--type <api|webhooks|db>",
        description: "Filter by operation type"
      }
    ];
  }
  async action({ options }) {
    const { plugin, type } = options;
    const instance = await getCorsairInstance({ cwd: process.cwd() });
    const result = listOperations(instance, {
      plugin,
      type
    });
    if (type === "db") {
      console.log(
        "[#corsair]: NOTE: Every DB query listed here has both .search() and .list() methods available.\n"
      );
    }
    console.log(result);
  }
};

// src/commands/schema.command.ts
init_corsair_instance();
import { getSchema } from "corsair";
var SchemaCommand = class extends BaseCommand {
  getName() {
    return "schema";
  }
  getDescription() {
    return "Show schema for a specific path";
  }
  getArguments() {
    return [{ name: "<path>", description: "Endpoint/webhook/db path" }];
  }
  async action({ args }) {
    const schemaPath = args[0];
    if (!schemaPath) {
      console.error("[#corsair]: Usage: corsair schema <path>");
      process.exit(1);
    }
    const instance = await getCorsairInstance({ cwd: process.cwd() });
    const result = getSchema(instance, schemaPath);
    console.log(result);
  }
};

// src/commands/script.command.ts
init_corsair_instance();
var AsyncFunction = Object.getPrototypeOf(async function() {
}).constructor;
var ScriptCommand = class extends BaseCommand {
  getName() {
    return "script";
  }
  getDescription() {
    return "Execute inline JavaScript with corsair injected";
  }
  getOptions() {
    return [
      {
        short: "-c",
        long: "--code <js>",
        description: "Async JavaScript body to execute"
      },
      {
        short: "-t",
        long: "--tenant <id>",
        description: "Tenant id for multi-tenant instances"
      }
    ];
  }
  async action({ options }) {
    const { code, tenant } = options;
    if (!code) {
      console.error('[#corsair]: Usage: corsair script --code "<js>"');
      process.exit(1);
    }
    let fn;
    try {
      fn = new AsyncFunction("corsair", code);
    } catch (error) {
      this.exitWithScriptError(error);
    }
    const instance = await getCorsairInstance({ cwd: process.cwd() });
    const client = resolveClient(instance, tenant);
    try {
      const result = await fn(client);
      if (result !== void 0) console.log(JSON.stringify(result, null, 2));
    } catch (error) {
      this.exitWithScriptError(error);
    }
  }
  exitWithScriptError(error) {
    const msg = error instanceof Error ? error.message : String(error);
    console.error(`[#corsair]: ${msg.slice(0, 500)}`);
    process.exit(1);
  }
};

// src/commands/setup.command.ts
import { setupCorsair } from "corsair";

// src/lib/setup-credentials.ts
var SETUP_FLAGS_WITH_VALUE = /* @__PURE__ */ new Set(["--tenant", "-t", "--plugin", "-p"]);
function parseSetupCredentials(rawArgs, pluginIds) {
  const pluginSet = new Set(pluginIds);
  const result = {};
  let currentPlugin = null;
  for (let i = 0; i < rawArgs.length; i++) {
    const arg = rawArgs[i];
    if (arg === "--backfill" || arg === "-b" || arg === "-backfill") {
      continue;
    }
    if (SETUP_FLAGS_WITH_VALUE.has(arg)) {
      const value2 = rawArgs[++i];
      if (arg === "--plugin" || arg === "-p") {
        if (value2 && pluginSet.has(value2)) {
          currentPlugin = value2;
          result[value2] ??= {};
        }
      }
      continue;
    }
    if (arg.startsWith("--tenant=") || arg.startsWith("-t=")) {
      continue;
    }
    if (arg.startsWith("--plugin=")) {
      const pluginId = arg.slice("--plugin=".length);
      if (pluginSet.has(pluginId)) {
        currentPlugin = pluginId;
        result[pluginId] ??= {};
      }
      continue;
    }
    if (arg.startsWith("--")) {
      const flagBody = arg.slice(2);
      const eqIdx2 = flagBody.indexOf("=");
      if (eqIdx2 > 0) {
        const name = flagBody.slice(0, eqIdx2);
        const value2 = flagBody.slice(eqIdx2 + 1);
        if (name === "plugin" && pluginSet.has(value2)) {
          currentPlugin = value2;
          result[value2] ??= {};
        }
        continue;
      }
      if (pluginSet.has(flagBody)) {
        currentPlugin = flagBody;
        result[flagBody] ??= {};
      }
      continue;
    }
    const eqIdx = arg.indexOf("=");
    if (eqIdx <= 0) continue;
    const field = arg.slice(0, eqIdx);
    const value = arg.slice(eqIdx + 1);
    if (!field || !value || !currentPlugin) continue;
    const bucket = result[currentPlugin] ??= {};
    bucket[field] = value;
  }
  return result;
}
function getSetupRawArgs(argv = process.argv) {
  const setupIdx = argv.indexOf("setup");
  return setupIdx >= 0 ? argv.slice(setupIdx + 1) : [];
}

// src/commands/setup.command.ts
init_corsair();
init_corsair_instance();
var SetupCommand = class extends BaseCommand {
  allowUnknownOptions() {
    return true;
  }
  getName() {
    return "setup";
  }
  getDescription() {
    return "Initialize Corsair and set plugin credentials";
  }
  getOptions() {
    return [
      {
        short: "-b",
        long: "--backfill",
        description: "Seed data while initializing"
      },
      {
        short: "-t",
        long: "--tenant <id>",
        description: "Tenant id for multi-tenant setup"
      },
      {
        short: "-p",
        long: "--plugin <id>",
        description: "Plugin id for credentials"
      }
    ];
  }
  getArguments() {
    return [
      {
        name: "[credentials...]",
        description: "Credential pairs like key=value"
      }
    ];
  }
  async action({ options }) {
    const cwd = process.cwd();
    const backfill = options.backfill || this.hasLegacyBackfillFlag();
    const tenantId = options.tenant;
    const instance = await getCorsairInstance({ cwd });
    const internal = await extractInternalConfig(cwd);
    const credentials = parseSetupCredentials(
      getSetupRawArgs(),
      internal.plugins.map((p13) => p13.id)
    );
    await setupCorsair(instance, {
      backfill,
      tenantId,
      credentials,
      caller: "cli"
    });
  }
  hasLegacyBackfillFlag() {
    const setupIdx = process.argv.indexOf("setup");
    const rawArgs = setupIdx >= 0 ? process.argv.slice(setupIdx + 1) : [];
    return rawArgs.some((arg) => arg === "-backfill");
  }
};

// src/commands/studio.command.ts
import path4 from "path";
var StudioCommand = class extends BaseCommand {
  getName() {
    return "ui";
  }
  getAliases() {
    return ["studio"];
  }
  getDescription() {
    return "Start Corsair Studio";
  }
  getOptions() {
    return [
      {
        short: "-p",
        long: "--port <number>",
        description: "Port to start Studio on"
      },
      { short: "-o", long: "--open", description: "Open browser on launch" },
      {
        short: "-O",
        long: "--no-open",
        description: "Do not open browser on launch"
      }
    ];
  }
  async action({ options }) {
    const cwd = process.cwd();
    const port = options.port ? Number.parseInt(options.port, 10) : void 0;
    let startStudio;
    try {
      const { createRequire } = await import("module");
      const { pathToFileURL } = await import("url");
      const req = createRequire(path4.join(cwd, "package.json"));
      const resolvedPath = req.resolve("@corsair-dev/studio/server");
      const mod = await import(pathToFileURL(resolvedPath).href);
      const candidate = mod.start ?? mod.startStudio;
      if (!candidate) {
        throw new Error("@corsair-dev/studio/server did not export `start`.");
      }
      startStudio = candidate;
    } catch (error) {
      const msg = error instanceof Error ? error.message : String(error);
      if (msg.includes("Cannot find package") || msg.includes("Cannot find module") || msg.includes("ERR_MODULE_NOT_FOUND") || msg.includes("MODULE_NOT_FOUND")) {
        console.error("[#corsair]: Corsair Studio is not installed.");
        console.error(
          "[#corsair]: Install it with: pnpm add -D @corsair-dev/studio"
        );
        process.exit(1);
      }
      throw error;
    }
    await startStudio({ cwd, port, open: options.open });
    await new Promise((resolve) => {
      const shutdown = () => resolve();
      process.once("SIGINT", shutdown);
      process.once("SIGTERM", shutdown);
    });
  }
};

// src/commands/subscribe/teams.command.ts
var TeamsCommand = class extends BaseCommand {
  getName() {
    return "teams";
  }
  getDescription() {
    return "Subscribe Teams webhooks";
  }
  async action({}) {
    const { runTeamsSubscribe: runTeamsSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
    await runTeamsSubscribe2({ cwd: process.cwd() });
  }
};

// src/commands/subscribe/sharepoint.command.ts
init_sharepoint();
var SharepointCommand = class extends BaseCommand {
  getName() {
    return "sharepoint";
  }
  getDescription() {
    return "Subscribe SharePoint webhooks";
  }
  async action({}) {
    await runSharepointSubscribe({ cwd: process.cwd() });
  }
};

// src/commands/subscribe/outlook.command.ts
init_subscribe_microsoft();
var OutlookCommand = class extends BaseCommand {
  getName() {
    return "outlook";
  }
  getDescription() {
    return "Subscribe Outlook webhooks";
  }
  async action({}) {
    await runOutlookSubscribe({ cwd: process.cwd() });
  }
};

// src/commands/subscribe/onedrive.command.ts
var OnedriveCommand = class extends BaseCommand {
  getName() {
    return "onedrive";
  }
  getDescription() {
    return "Subscribe OneDrive webhooks";
  }
  async action({}) {
    const { runOnedriveSubscribe: runOnedriveSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
    await runOnedriveSubscribe2({ cwd: process.cwd() });
  }
};

// src/commands/subscribe.command.ts
var SubscribeCommand = class extends BaseCommand {
  getName() {
    return "subscribe";
  }
  getDescription() {
    return "Subscribe to various kinds of webhooks";
  }
  getOptions() {
    return [
      {
        short: "-p",
        long: "--plugin <id>",
        description: "Plugin id (legacy: prefer `corsair subscribe <plugin>`)"
      }
    ];
  }
  getSubCommands() {
    return [
      new SharepointCommand(),
      new TeamsCommand(),
      new OutlookCommand(),
      new OnedriveCommand()
    ];
  }
  async action({ options }) {
    if (!options.plugin) {
      console.error(
        "Usage: corsair subscribe --plugin=<id> or corsair subscribe <plugin>"
      );
      console.error(
        "[#corsair]: Supported: outlook, sharepoint, teams, onedrive"
      );
      process.exit(1);
    }
    if (options.plugin === "outlook") {
      const { runOutlookSubscribe: runOutlookSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
      await runOutlookSubscribe2({ cwd: process.cwd() });
      return;
    }
    if (options.plugin === "sharepoint") {
      const { runSharepointSubscribe: runSharepointSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
      await runSharepointSubscribe2({ cwd: process.cwd() });
      return;
    }
    if (options.plugin === "teams") {
      const { runTeamsSubscribe: runTeamsSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
      await runTeamsSubscribe2({ cwd: process.cwd() });
      return;
    }
    if (options.plugin === "onedrive") {
      const { runOnedriveSubscribe: runOnedriveSubscribe2 } = await Promise.resolve().then(() => (init_subscribe_microsoft(), subscribe_microsoft_exports));
      await runOnedriveSubscribe2({ cwd: process.cwd() });
      return;
    }
    console.error(
      `[#corsair]: Unknown plugin for subscribe: '${options.plugin}'. Supported: outlook, sharepoint, teams, onedrive`
    );
    process.exit(1);
  }
};

// src/commands/sharepoint-subscribe.command.ts
var SharepointSubscribeCommand = class extends SharepointCommand {
  getName() {
    return "sharepoint-subscribe";
  }
  getDescription() {
    return "Subscribe SharePoint webhooks";
  }
};

// src/commands/teams-subscribe.command.ts
var TeamsSubscribeCommand = class extends TeamsCommand {
  getName() {
    return "teams-subscribe";
  }
  getDescription() {
    return "Subscribe Teams webhooks";
  }
};

// src/commands/onedrive-subscribe.command.ts
var OnedriveSubscribeCommand = class extends OnedriveCommand {
  getName() {
    return "onedrive-subscribe";
  }
  getDescription() {
    return "Subscribe OneDrive webhooks";
  }
};

// src/commands/watch-renew.command.ts
init_google();
var WatchRenewCommand = class extends BaseCommand {
  getName() {
    return "watch-renew";
  }
  getDescription() {
    return "Renew Google webhook subscriptions";
  }
  async action({}) {
    await runGoogleSubscribe({ cwd: process.cwd() });
  }
};

// src/index.ts
init_corsair_instance();
function createProgram() {
  const program = new Command();
  program.name("corsair");
  program.version(version, "-v, --version", "Output the current version");
  const COMMANDS = [
    new SetupCommand(),
    new AuthCommand(),
    new WatchRenewCommand(),
    new SubscribeCommand(),
    new SharepointSubscribeCommand(),
    new TeamsSubscribeCommand(),
    new OnedriveSubscribeCommand(),
    new ListCommand(),
    new SchemaCommand(),
    new ScriptCommand(),
    new StudioCommand()
  ];
  COMMANDS.forEach((command) => {
    command.prepare(program);
  });
  return program;
}
function normalizeLegacyArgs(argv) {
  return argv.map((arg, index) => {
    if (arg === "-backfill" && argv.slice(0, index).includes("setup")) {
      return "--backfill";
    }
    return arg;
  });
}
function detectIsMainModule() {
  const argv1 = process.argv[1];
  if (!argv1) return false;
  try {
    const argvResolved = realpathSync(argv1);
    const selfResolved = realpathSync(fileURLToPath(import.meta.url));
    return argvResolved === selfResolved;
  } catch {
    return import.meta.url === `file://${argv1.replace(/\\/g, "/")}`;
  }
}
if (detectIsMainModule()) {
  createProgram().parse(normalizeLegacyArgs(process.argv));
}
export {
  createProgram,
  findCorsairConfigPath,
  getCorsairInstance
};
